// @react-admin/ra-audit-log/src/Timeline/RecordTimelineItem.tsx
import * as React from 'react';
import { ReactElement } from 'react';
import { ListItem, ListItemText, Typography } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { ClassesOverride, useRecordContext, useTranslate } from 'react-admin';
import classnames from 'classnames';
import { EventRecord } from '../types';
import { EventAvatar } from '../EventAvatar';
import { useEventLabel } from '../useEventLabel';
import { TimelineItemProps } from './TimelineItem';

/**
 * Default component to display an audit logs.
 * @see Timeline
 */
export const RecordTimelineItem = (
    props: RecordTimelineItemProps
): ReactElement => {
    const { className, classes: classesOverride, ...rest } = props;
    const classes = useStyles(props);
    const translate = useTranslate();
    const record = useRecordContext<EventRecord>(props.record);
    const actionLabel = useEventLabel({ record, variant: 'record' });

    if (!record) {
        return null;
    }

    return (
        <ListItem
            className={classnames(classes.root, className)}
            button={false}
            dense
            disableGutters
            {...rest}
        >
            <ListItemText
                primary={
                    <div className={classes.authorContainer}>
                        <EventAvatar
                            className={classes.avatar}
                            alt={record.author.fullName}
                            src={record.author.avatar}
                            fullName={record.author.fullName}
                            role="presentation"
                            size="small"
                        />
                        <Typography
                            color="textPrimary"
                            className={classes.content}
                        >
                            <strong className={classes.author}>
                                {record.author.fullName ||
                                    translate(
                                        `ra-audit-log.author`,
                                        record.author
                                    )}
                            </strong>
                        </Typography>
                    </div>
                }
                secondary={
                    <span className={classes.action}>{actionLabel}</span>
                }
            />
        </ListItem>
    );
};

const useStyles = makeStyles(
    theme => ({
        root: {},
        content: {
            display: 'flex',
            flexWrap: 'wrap',
        },
        authorContainer: {
            display: 'flex',
            alignItems: 'center',
        },
        avatar: {
            marginRight: theme.spacing(1),
        },
        author: {},
        action: {},
        date: {},
    }),
    {
        name: 'RaTimelineItem',
    }
);

export interface RecordTimelineItemProps
    extends Omit<TimelineItemProps, 'link' | 'button'> {
    className?: string;
    classes?: ClassesOverride<typeof useStyles>;
    record?: EventRecord;
}
