// @react-admin/ra-navigation/lib/stories/MenuItem.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Long = exports.WithIconsDarkMode = exports.WithIcons = exports.InitiallyOpen = exports.OpenedOnLoad = exports.DarkMode = exports.Basic = void 0;
/* eslint-disable @typescript-eslint/explicit-function-return-type */
var React = __importStar(require("react"));
var react_1 = require("react");
var react_admin_1 = require("react-admin");
var history_1 = require("history");
var Card_1 = __importDefault(require("@material-ui/core/Card"));
var Dashboard_1 = __importDefault(require("@material-ui/icons/Dashboard"));
var MusicNote_1 = __importDefault(require("@material-ui/icons/MusicNote"));
var People_1 = __importDefault(require("@material-ui/icons/People"));
var merge_1 = __importDefault(require("lodash/merge"));
var app_location_1 = require("../src/app-location");
var multi_level_menu_1 = require("../src/multi-level-menu");
var dataProvider_1 = __importDefault(require("./dataProvider"));
exports.default = { title: 'ra-navigation/MultiLevelMenu/MenuItem' };
var BasicMultiLevelMenu = function () { return (React.createElement(multi_level_menu_1.MultiLevelMenu, null,
    React.createElement(multi_level_menu_1.MenuItem, { name: "dashboard", to: "/", exact: true, label: "Dashboard" }),
    React.createElement(multi_level_menu_1.MenuItem, { name: "songs", to: "/songs", label: "Songs" }),
    React.createElement(multi_level_menu_1.MenuItem, { name: "artists", to: '/artists?filter={}', label: "Artists" },
        React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock", to: '/artists?filter={"type":"Rock"}', label: "Rock" },
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock.pop", to: '/artists?filter={"type":"Pop Rock"}', label: "Pop Rock" }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock.folk", to: '/artists?filter={"type":"Folk Rock"}', label: "Folk Rock" })),
        React.createElement(multi_level_menu_1.MenuItem, { name: "artists.jazz", to: '/artists?filter={"type":"Jazz"}', label: "Jazz" },
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.jazz.rb", to: '/artists?filter={"type":"RB"}', label: "R&B" }))))); };
var BasicLayout = function (props) {
    return (React.createElement(app_location_1.AppLocationContext, null,
        React.createElement(react_admin_1.Layout, __assign({}, props, { menu: BasicMultiLevelMenu }))));
};
var Dashboard = function () {
    app_location_1.useDefineAppLocation('dashboard');
    return (React.createElement(Card_1.default, null,
        React.createElement("h1", null, "Dashboard")));
};
var SongList = function (props) { return (React.createElement(react_admin_1.List, __assign({}, props),
    React.createElement(react_admin_1.Datagrid, null,
        React.createElement(react_admin_1.TextField, { source: "id" }),
        React.createElement(react_admin_1.TextField, { source: "title" }),
        React.createElement(react_admin_1.DateField, { source: "released" })))); };
var ArtistList = function (props) { return (React.createElement(react_admin_1.List, __assign({}, props),
    React.createElement(ArtistsDatagrid, null))); };
var types = {
    Rock: 'artists.rock',
    'Folk Rock': 'artists.rock.folk',
    'Pop Rock': 'artists.rock.pop',
    Jazz: 'artists.jazz',
    RB: 'artists.jazz.rb',
};
var ArtistsDatagrid = function (props) {
    var _a = app_location_1.useAppLocationState(), setLocation = _a[1];
    var resourceLocation = app_location_1.useResourceAppLocation();
    react_1.useEffect(function () {
        var type = props.filterValues.type;
        if (typeof type !== 'undefined') {
            setLocation(types[type]);
        }
    }, 
    /* eslint-disable react-hooks/exhaustive-deps */
    [
        JSON.stringify({
            resourceLocation: resourceLocation,
            filter: props.filterValues,
        }),
    ]
    /* eslint-enable react-hooks/exhaustive-deps */
    );
    return (React.createElement(react_admin_1.Datagrid, __assign({}, props),
        React.createElement(react_admin_1.TextField, { source: "id" }),
        React.createElement(react_admin_1.TextField, { source: "name" })));
};
var Basic = function () { return (React.createElement(react_admin_1.Admin, { history: history_1.createMemoryHistory(), dataProvider: dataProvider_1.default, layout: BasicLayout, dashboard: Dashboard },
    React.createElement(react_admin_1.Resource, { name: "songs", list: SongList }),
    React.createElement(react_admin_1.Resource, { name: "artists", list: ArtistList }))); };
exports.Basic = Basic;
var DarkMode = function () {
    var darkTheme = merge_1.default({}, multi_level_menu_1.theme, {
        palette: {
            type: 'dark',
        },
    });
    return (React.createElement(react_admin_1.Admin, { history: history_1.createMemoryHistory(), dataProvider: dataProvider_1.default, layout: BasicLayout, dashboard: Dashboard, theme: darkTheme },
        React.createElement(react_admin_1.Resource, { name: "songs", list: SongList }),
        React.createElement(react_admin_1.Resource, { name: "artists", list: ArtistList })));
};
exports.DarkMode = DarkMode;
var OpenedOnLoad = function () {
    var history = history_1.createMemoryHistory({
        initialEntries: ['/artists?filter={"type":"Folk Rock"}'],
    });
    return (React.createElement(react_admin_1.Admin, { dataProvider: dataProvider_1.default, layout: BasicLayout, dashboard: Dashboard, history: history },
        React.createElement(react_admin_1.Resource, { name: "songs", list: SongList }),
        React.createElement(react_admin_1.Resource, { name: "artists", list: ArtistList })));
};
exports.OpenedOnLoad = OpenedOnLoad;
var InitiallyOpenMultiLevelMenu = function () { return (React.createElement(multi_level_menu_1.MultiLevelMenu, { initialOpen: true },
    React.createElement(multi_level_menu_1.MenuItem, { name: "dashboard", to: "/", exact: true, label: "Dashboard" }),
    React.createElement(multi_level_menu_1.MenuItem, { name: "songs", to: "/songs", label: "Songs" }),
    React.createElement(multi_level_menu_1.MenuItem, { name: "artists", to: '/artists?filter={}', label: "Artists" },
        React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock", to: '/artists?filter={"type":"Rock"}', label: "Rock" },
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock.pop", to: '/artists?filter={"type":"Pop Rock"}', label: "Pop Rock" }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock.folk", to: '/artists?filter={"type":"Folk Rock"}', label: "Folk Rock" })),
        React.createElement(multi_level_menu_1.MenuItem, { name: "artists.jazz", to: '/artists?filter={"type":"Jazz"}', label: "Jazz" },
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.jazz.rb", to: '/artists?filter={"type":"RB"}', label: "R&B" }))))); };
var InitiallyOpenLayout = function (props) {
    return (React.createElement(app_location_1.AppLocationContext, null,
        React.createElement(react_admin_1.Layout, __assign({}, props, { menu: InitiallyOpenMultiLevelMenu }))));
};
var InitiallyOpen = function () { return (React.createElement(react_admin_1.Admin, { history: history_1.createMemoryHistory(), dataProvider: dataProvider_1.default, layout: InitiallyOpenLayout, dashboard: Dashboard },
    React.createElement(react_admin_1.Resource, { name: "songs", list: SongList }),
    React.createElement(react_admin_1.Resource, { name: "artists", list: ArtistList }))); };
exports.InitiallyOpen = InitiallyOpen;
var MultiLevelMenuWithIcons = function () { return (React.createElement(multi_level_menu_1.MultiLevelMenu, null,
    React.createElement(multi_level_menu_1.MenuItem, { name: "dashboard", to: "/", exact: true, label: "Dashboard", icon: React.createElement(Dashboard_1.default, null) }),
    React.createElement(multi_level_menu_1.MenuItem, { name: "songs", to: "/songs", label: "Songs", icon: React.createElement(MusicNote_1.default, null) }),
    React.createElement(multi_level_menu_1.MenuItem, { name: "artists", to: '/artists?filter={}', label: "Artists", icon: React.createElement(People_1.default, null) },
        React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock", to: '/artists?filter={"type":"Rock"}', label: "Rock" },
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock.pop", to: '/artists?filter={"type":"Pop Rock"}', label: "Pop Rock" }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock.folk", to: '/artists?filter={"type":"Folk Rock"}', label: "Folk Rock" })),
        React.createElement(multi_level_menu_1.MenuItem, { name: "artists.jazz", to: '/artists?filter={"type":"Jazz"}', label: "Jazz" },
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.jazz.rb", to: '/artists?filter={"type":"RB"}', label: "R&B" }))))); };
var LayoutWithIcons = function (props) {
    return (React.createElement(app_location_1.AppLocationContext, null,
        React.createElement(react_admin_1.Layout, __assign({}, props, { menu: MultiLevelMenuWithIcons }))));
};
var WithIcons = function () { return (React.createElement(react_admin_1.Admin, { dataProvider: dataProvider_1.default, layout: LayoutWithIcons, dashboard: Dashboard, history: history_1.createMemoryHistory() },
    React.createElement(react_admin_1.Resource, { name: "songs", list: SongList }),
    React.createElement(react_admin_1.Resource, { name: "artists", list: ArtistList }))); };
exports.WithIcons = WithIcons;
var WithIconsDarkMode = function () {
    var darkTheme = merge_1.default({}, multi_level_menu_1.theme, {
        palette: {
            type: 'dark',
        },
    });
    return (React.createElement(react_admin_1.Admin, { history: history_1.createMemoryHistory(), dataProvider: dataProvider_1.default, layout: LayoutWithIcons, dashboard: Dashboard, theme: darkTheme },
        React.createElement(react_admin_1.Resource, { name: "songs", list: SongList }),
        React.createElement(react_admin_1.Resource, { name: "artists", list: ArtistList })));
};
exports.WithIconsDarkMode = WithIconsDarkMode;
var LongMultiLevelMenu = function () { return (React.createElement(multi_level_menu_1.MultiLevelMenu, { initialOpen: true },
    React.createElement(multi_level_menu_1.MenuItem, { name: "dashboard", to: "/", exact: true, label: "Dashboard", icon: React.createElement(People_1.default, null) }),
    React.createElement(multi_level_menu_1.MenuItem, { name: "songs", to: '/songs?filter={"released_gte":undefined,"released_lte":undefined}', label: "Songs", icon: React.createElement(People_1.default, null) },
        React.createElement(multi_level_menu_1.MenuItem, { name: "songs.60s", to: '/songs?filter={"released_gte":"01/01/1960","released_lte":"31/12/1969"}', label: "60s", icon: React.createElement(People_1.default, null) },
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.60s.pop", to: '/songs?filter={"type":"Pop Rock"}', label: "Pop Rock", icon: React.createElement(People_1.default, null) }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.60s.folk", to: '/songs?filter={"type":"Folk Rock"}', label: "Folk Rock", icon: React.createElement(People_1.default, null) }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.60s.rb", to: '/songs?filter={"type":"RB"}', label: "R&B", icon: React.createElement(People_1.default, null) })),
        React.createElement(multi_level_menu_1.MenuItem, { name: "songs.70s", to: '/songs?filter={"released_gte":"01/01/1970","released_lte":"31/12/1979"}', label: "70s", icon: React.createElement(People_1.default, null) },
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.70s.pop", to: '/songs?filter={"type":"Pop Rock"}', label: "Pop Rock", icon: React.createElement(People_1.default, null) }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.70s.folk", to: '/songs?filter={"type":"Folk Rock"}', label: "Folk Rock", icon: React.createElement(People_1.default, null) }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.70s.rb", to: '/songs?filter={"type":"RB"}', label: "R&B", icon: React.createElement(People_1.default, null) })),
        React.createElement(multi_level_menu_1.MenuItem, { name: "songs.80s", to: '/songs?filter={"released_gte":"01/01/1980","released_lte":"31/12/1989"}', label: "80s", icon: React.createElement(People_1.default, null) },
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.80s.pop", to: '/songs?filter={"type":"Pop Rock"}', label: "Pop Rock", icon: React.createElement(People_1.default, null) }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.80s.folk", to: '/songs?filter={"type":"Folk Rock"}', label: "Folk Rock", icon: React.createElement(People_1.default, null) }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.80s.rb", to: '/songs?filter={"type":"RB"}', label: "R&B", icon: React.createElement(People_1.default, null) })),
        React.createElement(multi_level_menu_1.MenuItem, { name: "songs.90s", to: '/songs?filter={"released_gte":"01/01/1990","released_lte":"31/12/1999"}', label: "90s", icon: React.createElement(People_1.default, null) },
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.90s.pop", to: '/songs?filter={"type":"Pop Rock"}', label: "Pop Rock", icon: React.createElement(People_1.default, null) }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.90s.folk", to: '/songs?filter={"type":"Folk Rock"}', label: "Folk Rock", icon: React.createElement(People_1.default, null) }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "songs.90s.rb", to: '/songs?filter={"type":"RB"}', label: "R&B", icon: React.createElement(People_1.default, null) }))),
    React.createElement(multi_level_menu_1.MenuItem, { name: "artists", to: '/artists?filter={}', label: "Artists", icon: React.createElement(People_1.default, null) },
        React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock", to: '/artists?filter={"type":"Rock"}', label: "Rock", icon: React.createElement(People_1.default, null) },
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock.pop", to: '/artists?filter={"type":"Pop Rock"}', label: "Pop Rock", icon: React.createElement(People_1.default, null) }),
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.rock.folk", to: '/artists?filter={"type":"Folk Rock"}', label: "Folk Rock", icon: React.createElement(People_1.default, null) })),
        React.createElement(multi_level_menu_1.MenuItem, { name: "artists.jazz", to: '/artists?filter={"type":"Jazz"}', label: "Jazz", icon: React.createElement(People_1.default, null) },
            React.createElement(multi_level_menu_1.MenuItem, { name: "artists.jazz.rb", to: '/artists?filter={"type":"RB"}', label: "R&B", icon: React.createElement(People_1.default, null) }))))); };
var LongLayout = function (props) {
    return (React.createElement(app_location_1.AppLocationContext, null,
        React.createElement(react_admin_1.Layout, __assign({}, props, { menu: LongMultiLevelMenu }))));
};
var Long = function () {
    return (React.createElement(react_admin_1.Admin, { history: history_1.createMemoryHistory(), dataProvider: dataProvider_1.default, layout: LongLayout, dashboard: Dashboard },
        React.createElement(react_admin_1.Resource, { name: "songs", list: SongList }),
        React.createElement(react_admin_1.Resource, { name: "artists", list: ArtistList })));
};
exports.Long = Long;
