// @react-admin/ra-markdown/lib/stories/MarkdownInput.stories.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithErrorDark = exports.WithError = exports.WithHelperTextDark = exports.WithHelperText = exports.SimpleDark = exports.Simple = void 0;
var react_1 = __importStar(require("react"));
var react_admin_1 = require("react-admin");
var ra_test_1 = require("ra-test");
var react_final_form_1 = require("react-final-form");
var core_1 = require("@material-ui/core");
var MarkdownInput_1 = require("../src/MarkdownInput");
exports.default = { title: 'ra-markdown/MarkdownInput' };
var MyToolbar = function () { return (react_1.default.createElement(core_1.Toolbar, null,
    react_1.default.createElement("span", null))); };
var record = { id: 0 };
var Simple = function () { return (react_1.default.createElement(ra_test_1.TestContext, { initialState: initialState },
    react_1.default.createElement(core_1.Card, null,
        react_1.default.createElement(core_1.CardContent, null,
            react_1.default.createElement(react_admin_1.SimpleForm, { resource: "posts", record: record, toolbar: react_1.default.createElement(MyToolbar, null) },
                react_1.default.createElement(MarkdownInput_1.MarkdownInput, { label: "Body", source: "body" })))))); };
exports.Simple = Simple;
var SimpleDark = function () { return (react_1.default.createElement(core_1.ThemeProvider, { theme: core_1.createMuiTheme({ palette: { type: 'dark' } }) },
    react_1.default.createElement(exports.Simple, null))); };
exports.SimpleDark = SimpleDark;
var WithHelperText = function () { return (react_1.default.createElement(ra_test_1.TestContext, { initialState: initialState },
    react_1.default.createElement(core_1.Card, null,
        react_1.default.createElement(core_1.CardContent, null,
            react_1.default.createElement(react_admin_1.SimpleForm, { resource: "posts", record: record, toolbar: react_1.default.createElement(MyToolbar, null) },
                react_1.default.createElement(MarkdownInput_1.MarkdownInput, { label: "Body", source: "body", helperText: "Enter your content here" })))))); };
exports.WithHelperText = WithHelperText;
var WithHelperTextDark = function () { return (react_1.default.createElement(core_1.ThemeProvider, { theme: core_1.createMuiTheme({ palette: { type: 'dark' } }) },
    react_1.default.createElement(exports.WithHelperText, null))); };
exports.WithHelperTextDark = WithHelperTextDark;
var validate = [react_admin_1.required()];
var ValidateOnLoadToolbar = function () {
    var form = react_final_form_1.useForm();
    react_1.useEffect(function () {
        form.focus('body');
        form.change('body', '');
        form.blur('body');
        form.submit();
    }, [form]);
    return react_1.default.createElement(MyToolbar, null);
};
var WithError = function () { return (react_1.default.createElement(ra_test_1.TestContext, { initialState: initialState },
    react_1.default.createElement(core_1.Card, null,
        react_1.default.createElement(core_1.CardContent, null,
            react_1.default.createElement(react_admin_1.SimpleForm, { resource: "posts", record: record, toolbar: react_1.default.createElement(ValidateOnLoadToolbar, null), save: function () {
                    /* */
                } },
                react_1.default.createElement(MarkdownInput_1.MarkdownInput, { label: "Body", source: "body", helperText: "Enter your content here", validate: validate })))))); };
exports.WithError = WithError;
var WithErrorDark = function () { return (react_1.default.createElement(core_1.ThemeProvider, { theme: core_1.createMuiTheme({ palette: { type: 'dark' } }) },
    react_1.default.createElement(exports.WithError, null))); };
exports.WithErrorDark = WithErrorDark;
var initialState = {
    admin: {
        resources: {
            posts: {
                data: { 7: { id: 7 } },
            },
        },
    },
};
