// @react-admin/ra-tour/src/Tour.tsx
import React from 'react';
import Joyride from 'react-joyride';
import { ErrorBoundary } from 'react-error-boundary';

import useTour from './useTour';
import Tooltip from './ui/Tooltip';
import { TourType } from './types';
import { useTranslate } from 'react-admin';

interface TourProps {
    joyrideProps?: any;
    tours: { [id: string]: TourType };
}

interface ErrorFallbackProps {
    error?: any;
    componentStack?: any;
    resetErrorBoundary?: any;
}

const ErrorFallback = ({
    error,
    componentStack,
    resetErrorBoundary,
}: ErrorFallbackProps) => {
    process.env.NODE_ENV !== 'production' &&
        console.warn('Joyride error', {
            error,
            componentStack,
            resetErrorBoundary,
        });
    return null;
};

/**
 * Component that handles the UI part of the tour.
 *
 * Uses react-joyride
 *
 * @param {TourType} tours List of supported tours
 * @param {any} joyrideProps Props passed down to react-joyride
 */
const Tour = ({ tours, joyrideProps }: TourProps) => {
    const [{ running, stepIndex, activeTour }, { stop }] = useTour();
    const translate = useTranslate();

    const defaultJoyRideProps = {
        locale: {
            back: translate('ra.action.back', { _: 'Back' }),
            close: translate('ra.action.close', { _: 'Close' }),
            last: translate('ra-tour.action.last', { _: 'Last' }),
            next: translate('ra-tour.action.next', { _: 'Next' }),
            open: translate('ra-tour.action.open', { _: 'Open the dialog' }),
            skip: translate('ra-tour.action.skip', { _: 'Skip' }),
        },
    };

    if (!activeTour) {
        return null;
    }

    const { steps } = tours[activeTour];

    if (!running) {
        return null;
    }

    const stepJoyrideProps = steps[stepIndex]
        ? steps[stepIndex].joyrideProps
        : {};

    return (
        <ErrorBoundary
            FallbackComponent={ErrorFallback}
            onError={(error: Error, componentStack: string): void => {
                process.env.NODE_ENV !== 'production' &&
                    console.warn('Joyride error', { error, componentStack });
            }}
            onReset={(): void => {
                stop();
            }}
        >
            <Joyride
                showProgress
                showSkipButton
                disableCloseOnEsc
                disableOverlayClose
                spotlightClicks
                tooltipComponent={Tooltip}
                {...defaultJoyRideProps}
                {...joyrideProps}
                {...stepJoyrideProps}
                steps={steps.map(step => ({
                    ...step,
                    content:
                        typeof step.content === 'string'
                            ? translate(step.content, {
                                  _: step.content,
                              })
                            : step.content,
                }))}
                run={true}
                stepIndex={stepIndex}
                continuous
            />
        </ErrorBoundary>
    );
};

export default Tour;
