// @react-admin/ra-audit-log/lib/src/EventAvatar.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInitials = exports.EventAvatar = void 0;
var React = __importStar(require("react"));
var core_1 = require("@material-ui/core");
var styles_1 = require("@material-ui/core/styles");
var Person_1 = __importDefault(require("@material-ui/icons/Person"));
var classnames_1 = __importDefault(require("classnames"));
/**
 * A Material-UI Avatar with fallback to the Person icon if no url was provided for an image.
 */
var EventAvatar = function (props) {
    var className = props.className, src = props.src, fullName = props.fullName, _a = props.size, size = _a === void 0 ? 'medium' : _a, rest = __rest(props, ["className", "src", "fullName", "size"]);
    var classes = useStyles(props);
    return (React.createElement(core_1.Avatar, __assign({ src: src, className: classnames_1.default(className, classes[size]) }, rest), !!src ? null : fullName ? exports.getInitials(fullName) : React.createElement(Person_1.default, null)));
};
exports.EventAvatar = EventAvatar;
var useStyles = styles_1.makeStyles({
    small: {
        width: 20,
        height: 20,
    },
    medium: {
        width: 40,
        height: 40,
    },
});
var getInitials = function (name) {
    return name
        .match(/(^\w\w?|\b\w)?/g)
        .join('')
        .match(/(^\w|\w$)?/g)
        .join('')
        .toUpperCase();
};
exports.getInitials = getInitials;
