// @react-admin/ra-editable-datagrid/src/buttons/SaveRowButton.tsx
import React from 'react';
import { getMutationMode, MutationMode, useTranslate } from 'react-admin';
import { Tooltip, IconButton } from '@material-ui/core';
import ContentSave from '@material-ui/icons/Save';

export const SaveRowButton = ({
    dirty,
    handleSubmit,
    invalid,
    mutationMode,
    quitEditMode,
    saving,
    undoable,
}: SaveRowButtonProps) => {
    const translate = useTranslate();
    const onClick = (
        evt: React.MouseEvent<HTMLButtonElement, MouseEvent>
    ): void => {
        evt.stopPropagation();
        evt.preventDefault();

        if (dirty && invalid) {
            return;
        }

        handleSubmit();
        const mode = getMutationMode(mutationMode, undoable);

        if (mode === 'undoable') {
            quitEditMode();
        }
    };

    const label = translate('ra.action.save', {
        _: 'ra.action.save',
    });

    return (
        <Tooltip title={label}>
            <IconButton
                aria-label={label}
                disabled={saving}
                onClick={onClick}
                size="small"
                color="primary"
            >
                <ContentSave />
            </IconButton>
        </Tooltip>
    );
};

export interface SaveRowButtonProps {
    dirty: boolean;
    handleSubmit: () => void;
    invalid: boolean;
    mutationMode?: MutationMode;
    quitEditMode?: () => void;
    saving?: boolean;
    undoable?: boolean;
}
