// @react-admin/ra-preferences/lib/src/preferencesStorage.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreferencesChanged = exports.overwriteStorage = exports.writeStorage = exports.readStorage = exports.getPreferencesStorage = void 0;
var set_1 = __importDefault(require("lodash/set"));
var get_1 = __importDefault(require("lodash/get"));
var constants_1 = require("./constants");
var getPreferencesStorage = function () {
    if (localStorageAvailable == undefined) {
        localStorageAvailable = testLocalStorage();
    }
    return localStorageAvailable ? window.localStorage : memoryStorage;
};
exports.getPreferencesStorage = getPreferencesStorage;
/**
 * reads key preference from storage
 *
 * @param {string} key Path of the preference key to read, uses lodash.get
 *
 * @return {T} Preference value

 * @example // Read full preferences
 *
 * const preferences = readStorage<PreferenceType>();
 *
 * @example // read preference for a particular key
 *
 * const density = readStorage<string>('density');

 * @example // read preference for a deep key
 *
 * const columns = readStorage<string[]>('orders.list.columns');
 */
var readStorage = function (key) {
    var preferences = JSON.parse(exports.getPreferencesStorage().getItem(constants_1.RA_PREFERENCE_WORKSPACE)) ||
        {};
    if (key) {
        return tryParse(get_1.default(preferences, key));
    }
    return (tryParse(preferences) || {});
};
exports.readStorage = readStorage;
/**
 * sets key preference to value on the preference storage
 *
 * @param {string} key Path of the preference key to update, uses lodash.set
 * @param {T} value New preference value
 *
 * @return {void}
 *
 * @example // Write full preferences to localstorage
 *
 * writeStorage<PreferenceType>('', {
 *      users: {
 *          lists: {
 *              columns: ['id', 'address'],
 *          },
 *      },
 *  });
 *
 * @example // Update preference for a particular key
 *
 * writeStorage<string>('density', 'small');
 *
 * @example // Update preference for a deep key
 *
 * writeStorage<string[]>('orders.list.columns', ['id', 'date', 'customer']);
 */
var writeStorage = function (key, value) {
    var storage = exports.getPreferencesStorage();
    try {
        var preferences = JSON.parse(storage.getItem(constants_1.RA_PREFERENCE_WORKSPACE)) || {};
        var newPreferences = key === '' ? value : set_1.default(__assign({}, preferences), key, value);
        storage.setItem(constants_1.RA_PREFERENCE_WORKSPACE, JSON.stringify(newPreferences));
        window.dispatchEvent(exports.PreferencesChanged.create(newPreferences));
        return newPreferences;
    }
    catch (err) {
        if (err instanceof TypeError &&
            err.message.includes('circular structure')) {
            throw new TypeError('The object that was given to the writeStorage function has circular references.\n' +
                'For more information, check here: ' +
                'https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cyclic_object_value');
        }
        throw err;
    }
};
exports.writeStorage = writeStorage;
var overwriteStorage = function (value) {
    var storage = exports.getPreferencesStorage();
    storage.setItem(constants_1.RA_PREFERENCE_WORKSPACE, JSON.stringify(value));
};
exports.overwriteStorage = overwriteStorage;
var localStorageAvailable = undefined;
var storage = {};
var memoryStorage = {
    getItem: function (key) {
        return get_1.default(storage, key);
    },
    setItem: function (key, value) {
        set_1.default(storage, key, value);
    },
};
var tryParse = function (value) {
    try {
        return JSON.parse(value);
    }
    catch (_a) {
        return value;
    }
};
var testLocalStorage = function () {
    if (window.localStorage == undefined) {
        return false;
    }
    try {
        window.localStorage.setItem('test', 'test');
        window.localStorage.removeItem('test');
        return true;
    }
    catch (_a) {
        return false;
    }
};
exports.PreferencesChanged = {
    eventName: 'onPreferencesChange',
    create: function (payload) {
        return new CustomEvent(exports.PreferencesChanged.eventName, {
            detail: payload,
        });
    },
};
