// @react-admin/ra-calendar/src/CompleteCalendar.tsx
import * as React from 'react';
import { ReactElement, FC } from 'react';
import { List, ListProps } from 'react-admin';
import {
    EditDialog,
    EditDialogProps,
    CreateDialog,
    CreateDialogProps,
} from '@react-admin/ra-form-layout';

import { Calendar, CalendarProps } from './Calendar';
import { getFilterValuesFromInterval } from './getFilterValuesFromInterval';

/**
 * Replace the List and Edit views by a Calendar, powered by Full Calendar
 *
 * This all-in one component renders a calendar, as well as a form to edit or
 * create new events that opens in a dialog.
 *
 * Use it as the `list` prop of a `<Resource>`. No need to specify an `edit`
 * or `create` prop for this resource, but you'll have to pass a form component
 * (like `<SimpleForm>`) as child of `<CompleteCalendar>` to define the event
 * edition form.
 *
 * Accepted props:
 * - children: A Form element with Input children
 * - ListProps: Props passed to the `<List>` component
 * - CalendarProps: Props passed to the `<Calendar>` component
 * - EditDialogProps: Props passed to the `<EditDialog>` component
 * - CreateDialogProps: Props passed to the `<CreateDialog>` component
 *
 * @example
 * ```
 * import React, { FC } from 'react';
 * import {
 *     Admin,
 *     Resource,
 *     List,
 *     ListProps,
 *     SimpleForm,
 *     TextInput,
 *     DateTimeInput,
 * } from 'react-admin';
 * import { CompleteCalendar } from '@react-admin/ra-calendar';
 *
 * import dataProvider from './dataProvider';
 *
 * const EventList: FC<ListProps> = props => (
 *     <CompleteCalendar {...props}>
 *         <SimpleForm>
 *             <TextInput source="title" autoFocus />
 *             <DateTimeInput source="start" />
 *             <DateTimeInput source="end" />
 *         </SimpleForm>
 *     </CompleteCalendar>
 * );
 *
 * export const Basic: FC = () => (
 *     <Admin dataProvider={dataProvider}>
 *         <Resource name="events" list={EventList} />
 *     </Admin>
 * );
 * ```
 */
export const CompleteCalendar: FC<CompleteCalendarProps> = props => {
    const {
        CalendarProps,
        EditDialogProps,
        CreateDialogProps,
        ListProps,
        children,
        ...rest
    } = props;
    return (
        <>
            <List
                filterDefaultValues={getFilterValuesFromInterval()}
                perPage={1000}
                {...ListProps}
                {...rest}
                pagination={false}
                hasCreate
            >
                <Calendar {...CalendarProps} />
            </List>
            <EditDialog title="Edit event" {...EditDialogProps} {...rest}>
                {children}
            </EditDialog>
            <CreateDialog {...CreateDialogProps} {...rest}>
                {children}
            </CreateDialog>
        </>
    );
};

export interface CompleteCalendarProps {
    children?: ReactElement;
    CalendarProps?: CalendarProps;
    ListProps?: ListProps;
    EditDialogProps?: EditDialogProps;
    CreateDialogProps?: CreateDialogProps;
}
