// @react-admin/ra-rbac/lib/stories/1-canAccess.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatagridColumns = void 0;
var react_1 = __importDefault(require("react"));
var react_admin_1 = require("react-admin");
var ra_data_fakerest_1 = __importDefault(require("ra-data-fakerest"));
var data_generator_retail_1 = __importDefault(require("data-generator-retail"));
var src_1 = require("../src");
exports.default = { title: 'ra-rbac/canAccess' };
var data = data_generator_retail_1.default();
var dataProvider = ra_data_fakerest_1.default(data, true);
var authProvider = {
    checkAuth: function () { return Promise.resolve(); },
    login: function () { return Promise.resolve(); },
    logout: function () { return Promise.resolve(); },
    checkError: function () { return Promise.resolve(); },
    getPermissions: function () {
        return promiseFor({
            permissions: [
                { action: 'list', resource: 'products' },
                { action: 'read', resource: 'products.price' },
            ],
        });
    },
};
var ProductList = function (_a) {
    var permissions = _a.permissions, props = __rest(_a, ["permissions"]);
    return (react_1.default.createElement(react_admin_1.List, __assign({}, props),
        react_1.default.createElement(react_admin_1.Datagrid, null,
            react_1.default.createElement(react_admin_1.TextField, { source: "id" }),
            react_1.default.createElement(react_admin_1.TextField, { source: "reference" }),
            react_1.default.createElement(react_admin_1.TextField, { source: "width" }),
            react_1.default.createElement(react_admin_1.TextField, { source: "height" }),
            src_1.canAccess({
                permissions: permissions,
                action: 'read',
                resource: 'products.price',
            }) && react_1.default.createElement(react_admin_1.TextField, { source: "price" }),
            src_1.canAccess({
                permissions: permissions,
                action: 'read',
                resource: 'products.stock',
            }) && react_1.default.createElement(react_admin_1.TextField, { source: "stock" }))));
};
var DatagridColumns = function () { return (react_1.default.createElement(react_admin_1.Admin, { authProvider: authProvider, dataProvider: dataProvider },
    react_1.default.createElement(src_1.Resource, { name: "products", list: ProductList }),
    react_1.default.createElement(src_1.Resource, { name: "categories" }))); };
exports.DatagridColumns = DatagridColumns;
var promiseFor = function (result) {
    return new Promise(function (resolve) {
        setTimeout(function () {
            resolve(result);
        }, 300);
    });
};
