// @react-admin/ra-rbac/lib/stories/3-WithPermissions.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdditionalProps = exports.AdminAccess = exports.NoAccess = exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var react_admin_1 = require("react-admin");
var src_1 = require("../src");
exports.default = { title: 'ra-rbac/WithPermissions' };
var authProvider = {
    checkAuth: function () { return Promise.resolve(); },
    login: function () { return Promise.resolve(); },
    logout: function () { return Promise.resolve(); },
    checkError: function () { return Promise.resolve(); },
    getPermissions: function () { return Promise.resolve(); },
    getRoles: function () { return Promise.resolve(); },
};
var PostRead = function () { return (react_1.default.createElement(src_1.WithPermissions, { action: "read", resource: "posts" }, "Allowed")); };
var PostWrite = function () { return (react_1.default.createElement(src_1.WithPermissions, { action: "write", resource: "posts" }, "Allowed")); };
var CommentRead = function () { return (react_1.default.createElement(src_1.WithPermissions, { action: "read", resource: "comments" }, "Allowed")); };
var App = function () { return (react_1.default.createElement("ul", null,
    react_1.default.createElement("li", null,
        "Read posts:",
        ' ',
        react_1.default.createElement("span", null,
            react_1.default.createElement(PostRead, null))),
    react_1.default.createElement("li", null,
        "Write posts:",
        ' ',
        react_1.default.createElement("span", null,
            react_1.default.createElement(PostWrite, null))),
    react_1.default.createElement("li", null,
        "Read comments:",
        ' ',
        react_1.default.createElement("span", null,
            react_1.default.createElement(CommentRead, null))))); };
var Basic = function () { return (react_1.default.createElement(react_admin_1.AuthContext.Provider, { value: __assign(__assign({}, authProvider), { getPermissions: function () {
            return promiseFor({
                permissions: [{ action: 'read', resource: 'posts' }],
                roles: ['commentReader'],
            });
        }, getRoles: function () {
            return promiseFor({
                commentReader: [{ action: 'read', resource: 'comments' }],
            });
        } }) },
    react_1.default.createElement(App, null))); };
exports.Basic = Basic;
var NoAccess = function () { return (react_1.default.createElement(react_admin_1.AuthContext.Provider, { value: __assign(__assign({}, authProvider), { getPermissions: function () { return promiseFor({ permissions: [] }); } }) },
    react_1.default.createElement(App, null))); };
exports.NoAccess = NoAccess;
var AdminAccess = function () { return (react_1.default.createElement(react_admin_1.AuthContext.Provider, { value: __assign(__assign({}, authProvider), { getPermissions: function () {
            return promiseFor({ permissions: [{ action: '*', resource: '*' }] });
        } }) },
    react_1.default.createElement(App, null))); };
exports.AdminAccess = AdminAccess;
var RenderProps = function (props) { return (react_1.default.createElement("ul", null, Object.keys(props).map(function (key) { return (react_1.default.createElement("li", { key: key }, props[key])); }))); };
var AdditionalProps = function () { return (react_1.default.createElement(react_admin_1.AuthContext.Provider, { value: __assign(__assign({}, authProvider), { getPermissions: function () {
            return promiseFor({ permissions: [{ action: '*', resource: '*' }] });
        } }) },
    react_1.default.createElement(src_1.WithPermissions, { action: "read", resource: "bar", name: "John", age: 34 },
        react_1.default.createElement(RenderProps, null)))); };
exports.AdditionalProps = AdditionalProps;
var promiseFor = function (result) {
    return new Promise(function (resolve) {
        setTimeout(function () {
            resolve(result);
        }, 200);
    });
};
