// @react-admin/ra-rbac/lib/stories/4-Resource.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var react_admin_1 = require("react-admin");
var ra_data_fakerest_1 = __importDefault(require("ra-data-fakerest"));
var data_generator_retail_1 = __importDefault(require("data-generator-retail"));
var src_1 = require("../src");
exports.default = { title: 'ra-rbac/Resource' };
var data = data_generator_retail_1.default();
var dataProvider = ra_data_fakerest_1.default(data, true);
var authProvider = {
    checkAuth: function () { return Promise.resolve(); },
    login: function () { return Promise.resolve(); },
    logout: function () { return Promise.resolve(); },
    checkError: function () { return Promise.resolve(); },
    getPermissions: function () {
        return promiseFor({
            permissions: [
                { action: 'list', resource: 'products' },
                { action: 'list', resource: 'categories' },
                { action: '*', resource: 'commands' },
                { action: ['list', 'edit'], resource: 'customers' },
            ],
        });
    },
};
var Basic = function () { return (react_1.default.createElement(react_admin_1.Admin, { authProvider: authProvider, dataProvider: dataProvider },
    react_1.default.createElement(src_1.Resource, { name: "products", list: react_admin_1.ListGuesser, show: react_admin_1.ShowGuesser, edit: react_admin_1.EditGuesser }),
    react_1.default.createElement(src_1.Resource, { name: "categories", list: react_admin_1.ListGuesser, show: react_admin_1.ShowGuesser, edit: react_admin_1.EditGuesser }),
    react_1.default.createElement(src_1.Resource, { name: "commands", list: react_admin_1.ListGuesser, show: react_admin_1.ShowGuesser, edit: react_admin_1.EditGuesser }),
    react_1.default.createElement(src_1.Resource, { name: "customers", list: react_admin_1.ListGuesser, show: react_admin_1.ShowGuesser, edit: react_admin_1.EditGuesser }),
    react_1.default.createElement(src_1.Resource, { name: "reviews", list: react_admin_1.ListGuesser, show: react_admin_1.ShowGuesser, edit: react_admin_1.EditGuesser }))); };
exports.Basic = Basic;
var promiseFor = function (result) {
    return new Promise(function (resolve) {
        setTimeout(function () {
            resolve(result);
        }, 300);
    });
};
