// @react-admin/ra-rbac/lib/stories/5-Menu.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var react_admin_1 = require("react-admin");
var ra_data_fakerest_1 = __importDefault(require("ra-data-fakerest"));
var data_generator_retail_1 = __importDefault(require("data-generator-retail"));
var src_1 = require("../src");
exports.default = { title: 'ra-rbac/Menu' };
var data = data_generator_retail_1.default();
var dataProvider = ra_data_fakerest_1.default(data, true);
var authProvider = {
    checkAuth: function () { return Promise.resolve(); },
    login: function () { return Promise.resolve(); },
    logout: function () { return Promise.resolve(); },
    checkError: function () { return Promise.resolve(); },
    getPermissions: function () {
        return promiseFor({
            permissions: [
                { action: 'list', resource: 'products' },
                { action: 'list', resource: 'categories' },
                { action: '*', resource: 'commands' },
                { action: ['list', 'edit'], resource: 'customers' },
            ],
        });
    },
};
var CustomLayout = function (props) { return react_1.default.createElement(react_admin_1.Layout, __assign({}, props, { menu: src_1.Menu })); };
var Basic = function () { return (react_1.default.createElement(react_admin_1.Admin, { authProvider: authProvider, dataProvider: dataProvider, layout: CustomLayout },
    react_1.default.createElement(react_admin_1.Resource, { name: "products", list: react_admin_1.ListGuesser }),
    react_1.default.createElement(react_admin_1.Resource, { name: "categories", list: react_admin_1.ListGuesser }),
    react_1.default.createElement(react_admin_1.Resource, { name: "commands", list: react_admin_1.ListGuesser }),
    react_1.default.createElement(react_admin_1.Resource, { name: "customers", list: react_admin_1.ListGuesser }),
    react_1.default.createElement(react_admin_1.Resource, { name: "reviews", list: react_admin_1.ListGuesser }))); };
exports.Basic = Basic;
var promiseFor = function (result) {
    return new Promise(function (resolve) {
        setTimeout(function () {
            resolve(result);
        }, 300);
    });
};
