// @react-admin/ra-rbac/lib/src/Resource.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Resource = void 0;
var React = __importStar(require("react"));
var react_1 = require("react");
var react_redux_1 = require("react-redux");
var react_router_dom_1 = require("react-router-dom");
var react_admin_1 = require("react-admin");
var omit_1 = __importDefault(require("lodash/omit"));
var usePermissions_1 = require("./usePermissions");
var canAccess_1 = require("./canAccess");
var useAuthenticated_1 = require("./useAuthenticated");
var defaultOptions = {};
var ResourceRegister = function (_a) {
    var name = _a.name, list = _a.list, create = _a.create, edit = _a.edit, show = _a.show, icon = _a.icon, _b = _a.options, options = _b === void 0 ? defaultOptions : _b;
    var dispatch = react_redux_1.useDispatch();
    react_1.useEffect(function () {
        dispatch(react_admin_1.registerResource({
            name: name,
            options: options,
            hasList: !!list,
            hasEdit: !!edit,
            hasShow: !!show,
            hasCreate: !!create,
            icon: icon,
        }));
        return function () {
            dispatch(react_admin_1.unregisterResource(name));
        };
    }, [dispatch, name, create, edit, icon, list, show, options]);
    return null;
};
var ResourceRoutes = function (_a) {
    var name = _a.name, match = _a.match, list = _a.list, create = _a.create, edit = _a.edit, show = _a.show, _b = _a.options, options = _b === void 0 ? defaultOptions : _b;
    var isRegistered = react_redux_1.useSelector(function (state) { return !!state.admin.resources[name]; });
    var authChecked = useAuthenticated_1.useAuthenticated().loaded;
    var _c = usePermissions_1.usePermissions(), permissionsLoaded = _c.loaded, permissions = _c.permissions;
    var basePath = match ? match.path : '';
    var resourceData = react_1.useMemo(function () { return ({
        resource: name,
        options: options,
        hasList: !!list,
        hasEdit: !!edit,
        hasShow: !!show,
        hasCreate: !!create,
    }); }, [name, options, create, edit, list, show]);
    // match tends to change even on the same route ; using memo to avoid an extra render
    return react_1.useMemo(function () {
        // if the registration hasn't finished, no need to render
        if (!isRegistered) {
            return null;
        }
        // if the authProvider hasn't returned, no need to render
        if (!authChecked || !permissionsLoaded) {
            return null;
        }
        return (React.createElement(react_admin_1.ResourceContextProvider, { value: name },
            React.createElement(react_router_dom_1.Switch, null,
                create &&
                    canAccess_1.canAccess({
                        permissions: permissions,
                        action: 'create',
                        resource: name,
                    }) && (React.createElement(react_router_dom_1.Route, { path: basePath + "/create", render: function (routeProps) {
                        return react_1.createElement(create, __assign(__assign({ basePath: basePath, permissions: permissions }, omit_1.default(routeProps, 'staticContext')), resourceData));
                    } })),
                show &&
                    canAccess_1.canAccess({
                        permissions: permissions,
                        action: 'show',
                        resource: name,
                    }) && (React.createElement(react_router_dom_1.Route, { path: basePath + "/:id/show", render: function (routeProps) {
                        return react_1.createElement(show, __assign(__assign({ basePath: basePath, id: decodeURIComponent(routeProps.match
                                .params.id), permissions: permissions }, omit_1.default(routeProps, 'staticContext')), resourceData));
                    } })),
                edit &&
                    canAccess_1.canAccess({
                        permissions: permissions,
                        action: 'edit',
                        resource: name,
                    }) && (React.createElement(react_router_dom_1.Route, { path: basePath + "/:id", render: function (routeProps) {
                        return react_1.createElement(edit, __assign(__assign({ basePath: basePath, id: decodeURIComponent(routeProps.match
                                .params.id), permissions: permissions }, omit_1.default(routeProps, 'staticContext')), resourceData));
                    } })),
                list &&
                    canAccess_1.canAccess({
                        permissions: permissions,
                        action: 'list',
                        resource: name,
                    }) && (React.createElement(react_router_dom_1.Route, { path: "" + basePath, render: function (routeProps) {
                        return react_1.createElement(list, __assign(__assign({ basePath: basePath, permissions: permissions, 
                            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                            // @ts-ignore
                            syncWithLocation: true }, omit_1.default(routeProps, 'staticContext')), resourceData));
                    } })))));
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [
        basePath,
        name,
        create,
        edit,
        list,
        show,
        options,
        isRegistered,
        authChecked,
        permissionsLoaded,
    ]);
};
var Resource = function (_a) {
    var _b = _a.intent, intent = _b === void 0 ? 'route' : _b, props = __rest(_a, ["intent"]);
    return intent === 'registration' ? (React.createElement(ResourceRegister, __assign({}, props))) : (React.createElement(ResourceRoutes, __assign({}, props)));
};
exports.Resource = Resource;
