// @react-admin/ra-rbac/esm/stories/0-fullApp.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React from 'react';
import { Admin, Layout, ListGuesser, ShowGuesser, EditGuesser, } from 'react-admin';
import fakeRestDataProvider from 'ra-data-fakerest';
import generateData from 'data-generator-retail';
import { Resource, Menu } from '../src';
import { LoginPage } from './fullApp';
import product from './fullApp/product';
export default { title: 'ra-rbac/Full App' };
var roles = {
    accountant: [
        { action: ['list', 'show'], resource: 'products' },
        { action: 'read', resource: 'products.*' },
        { type: 'deny', action: 'read', resource: 'products.description' },
        { action: 'list', resource: 'categories' },
        { action: 'read', resource: 'categories.*' },
        { action: ['list', 'show'], resource: 'customers' },
        { action: 'read', resource: 'customers.*' },
        { action: '*', resource: 'invoices' },
    ],
    contentEditor: [
        {
            action: ['list', 'create', 'edit', 'delete', 'export'],
            resource: 'products',
        },
        { action: 'read', resource: 'products.*' },
        { type: 'deny', action: 'read', resource: 'products.stock' },
        { type: 'deny', action: 'read', resource: 'products.sales' },
        { action: 'write', resource: 'products.*' },
        { type: 'deny', action: 'write', resource: 'products.stock' },
        { type: 'deny', action: 'write', resource: 'products.sales' },
        { action: 'list', resource: 'categories' },
        { action: ['list', 'edit'], resource: 'customers' },
        { action: ['list', 'edit'], resource: 'reviews' },
    ],
    stockManager: [
        { action: ['list', 'edit', 'export'], resource: 'products' },
        { action: 'read', resource: 'products.*' },
        {
            type: 'deny',
            action: 'read',
            resource: 'products.description',
        },
        { action: 'write', resource: 'products.stock' },
        { action: 'write', resource: 'products.sales' },
        { action: 'list', resource: 'categories' },
    ],
    administrator: [{ action: '*', resource: '*' }],
};
var data = generateData();
var dataProvider = fakeRestDataProvider(data, true);
var authProvider = {
    checkAuth: function () {
        var role = localStorage.getItem('role');
        return role ? Promise.resolve() : Promise.reject();
    },
    login: function (_a) {
        var role = _a.role;
        localStorage.setItem('role', role);
        return Promise.resolve();
    },
    logout: function () {
        localStorage.removeItem('role');
        return Promise.resolve();
    },
    checkError: function () { return Promise.resolve(); },
    getIdentity: function () {
        var role = localStorage.getItem('role');
        return Promise.resolve({
            id: role,
            fullName: {
                accountant: 'Accountant',
                contentEditor: 'Content Editor',
                stockManager: 'Stock manager',
                administrator: 'Administrator',
            }[role],
        });
    },
    getPermissions: function () {
        var role = localStorage.getItem('role');
        return Promise.resolve({
            roles: [role],
        });
    },
    getRoles: function () { return Promise.resolve(roles); },
};
var CustomLayout = function (props) { return React.createElement(Layout, __assign({}, props, { menu: Menu })); };
export var FullApp = function () { return (React.createElement(Admin, { loginPage: LoginPage, layout: CustomLayout, authProvider: authProvider, dataProvider: dataProvider },
    React.createElement(Resource, __assign({ name: "products" }, product)),
    React.createElement(Resource, { name: "categories", list: ListGuesser, show: ShowGuesser, edit: EditGuesser }),
    React.createElement(Resource, { name: "commands", list: ListGuesser, show: ShowGuesser, edit: EditGuesser }),
    React.createElement(Resource, { name: "invoices", list: ListGuesser, show: ShowGuesser, edit: EditGuesser }),
    React.createElement(Resource, { name: "customers", list: ListGuesser, show: ShowGuesser, edit: EditGuesser }),
    React.createElement(Resource, { name: "reviews", list: ListGuesser, show: ShowGuesser, edit: EditGuesser }))); };
