// @react-admin/ra-rbac/esm/stories/3-WithPermissions.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React from 'react';
import { AuthContext } from 'react-admin';
import { WithPermissions } from '../src';
export default { title: 'ra-rbac/WithPermissions' };
var authProvider = {
    checkAuth: function () { return Promise.resolve(); },
    login: function () { return Promise.resolve(); },
    logout: function () { return Promise.resolve(); },
    checkError: function () { return Promise.resolve(); },
    getPermissions: function () { return Promise.resolve(); },
    getRoles: function () { return Promise.resolve(); },
};
var PostRead = function () { return (React.createElement(WithPermissions, { action: "read", resource: "posts" }, "Allowed")); };
var PostWrite = function () { return (React.createElement(WithPermissions, { action: "write", resource: "posts" }, "Allowed")); };
var CommentRead = function () { return (React.createElement(WithPermissions, { action: "read", resource: "comments" }, "Allowed")); };
var App = function () { return (React.createElement("ul", null,
    React.createElement("li", null,
        "Read posts:",
        ' ',
        React.createElement("span", null,
            React.createElement(PostRead, null))),
    React.createElement("li", null,
        "Write posts:",
        ' ',
        React.createElement("span", null,
            React.createElement(PostWrite, null))),
    React.createElement("li", null,
        "Read comments:",
        ' ',
        React.createElement("span", null,
            React.createElement(CommentRead, null))))); };
export var Basic = function () { return (React.createElement(AuthContext.Provider, { value: __assign(__assign({}, authProvider), { getPermissions: function () {
            return promiseFor({
                permissions: [{ action: 'read', resource: 'posts' }],
                roles: ['commentReader'],
            });
        }, getRoles: function () {
            return promiseFor({
                commentReader: [{ action: 'read', resource: 'comments' }],
            });
        } }) },
    React.createElement(App, null))); };
export var NoAccess = function () { return (React.createElement(AuthContext.Provider, { value: __assign(__assign({}, authProvider), { getPermissions: function () { return promiseFor({ permissions: [] }); } }) },
    React.createElement(App, null))); };
export var AdminAccess = function () { return (React.createElement(AuthContext.Provider, { value: __assign(__assign({}, authProvider), { getPermissions: function () {
            return promiseFor({ permissions: [{ action: '*', resource: '*' }] });
        } }) },
    React.createElement(App, null))); };
var RenderProps = function (props) { return (React.createElement("ul", null, Object.keys(props).map(function (key) { return (React.createElement("li", { key: key }, props[key])); }))); };
export var AdditionalProps = function () { return (React.createElement(AuthContext.Provider, { value: __assign(__assign({}, authProvider), { getPermissions: function () {
            return promiseFor({ permissions: [{ action: '*', resource: '*' }] });
        } }) },
    React.createElement(WithPermissions, { action: "read", resource: "bar", name: "John", age: 34 },
        React.createElement(RenderProps, null)))); };
var promiseFor = function (result) {
    return new Promise(function (resolve) {
        setTimeout(function () {
            resolve(result);
        }, 200);
    });
};
