// @react-admin/ra-rbac/esm/stories/4-Resource.stories.js
import React from 'react';
import { Admin, ListGuesser, ShowGuesser, EditGuesser } from 'react-admin';
import fakeRestDataProvider from 'ra-data-fakerest';
import generateData from 'data-generator-retail';
import { Resource } from '../src';
export default { title: 'ra-rbac/Resource' };
var data = generateData();
var dataProvider = fakeRestDataProvider(data, true);
var authProvider = {
    checkAuth: function () { return Promise.resolve(); },
    login: function () { return Promise.resolve(); },
    logout: function () { return Promise.resolve(); },
    checkError: function () { return Promise.resolve(); },
    getPermissions: function () {
        return promiseFor({
            permissions: [
                { action: 'list', resource: 'products' },
                { action: 'list', resource: 'categories' },
                { action: '*', resource: 'commands' },
                { action: ['list', 'edit'], resource: 'customers' },
            ],
        });
    },
};
export var Basic = function () { return (React.createElement(Admin, { authProvider: authProvider, dataProvider: dataProvider },
    React.createElement(Resource, { name: "products", list: ListGuesser, show: ShowGuesser, edit: EditGuesser }),
    React.createElement(Resource, { name: "categories", list: ListGuesser, show: ShowGuesser, edit: EditGuesser }),
    React.createElement(Resource, { name: "commands", list: ListGuesser, show: ShowGuesser, edit: EditGuesser }),
    React.createElement(Resource, { name: "customers", list: ListGuesser, show: ShowGuesser, edit: EditGuesser }),
    React.createElement(Resource, { name: "reviews", list: ListGuesser, show: ShowGuesser, edit: EditGuesser }))); };
var promiseFor = function (result) {
    return new Promise(function (resolve) {
        setTimeout(function () {
            resolve(result);
        }, 300);
    });
};
