// @react-admin/ra-audit-log/esm/stories/RecordTimeLine.stories.js
import * as React from 'react';
import { Box, Container, ListItem, ListItemText } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { TranslationProvider, mergeTranslations, useRecordContext, DataProviderContext, convertLegacyDataProvider, createAdminStore, } from 'react-admin';
import polyglotI18nProvider from 'ra-i18n-polyglot';
import englishMessages from 'ra-language-english';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { raAuditLogLanguageEnglish, RecordTimeline, TimelineGroup, useEventLabel, } from '../src';
import { Router } from 'react-router-dom';
export default { title: 'ra-audit-log/RecordTimeline' };
var records = [
    {
        id: 1,
        resource: 'products',
        action: 'create',
        date: new Date(2020, 0, 3, 13, 14, 0),
        payload: {
            data: {
                id: 123,
                name: 'SkyLine Poster',
                category: 'cities',
                price: 20,
                stock: 10,
            },
        },
        author: {
            id: 123,
            fullName: 'Thomas A. Anderson',
            avatar: 'https://upload.wikimedia.org/wikipedia/en/c/c6/NeoTheMatrix.jpg',
        },
    },
    {
        id: 2,
        resource: 'products',
        action: 'update',
        date: new Date(2020, 0, 6, 9, 48, 0),
        payload: {
            id: 789,
            data: {
                id: 123,
                name: 'NewYork SkyLine Poster',
                category: 'cities',
                price: 20,
                stock: 10,
            },
            previousData: {
                id: 123,
                name: 'SkyLine Poster',
                category: 'cities',
                price: 20,
                stock: 10,
            },
        },
        author: {
            id: 789,
            fullName: 'Takeshi Kovacs',
            avatar: 'https://cdn.images.express.co.uk/img/dynamic/20/285x395/922644_1.jpg',
        },
    },
    {
        id: 3,
        resource: 'products',
        action: 'update',
        date: new Date(2020, 0, 6, 11, 48, 0),
        payload: {
            id: 789,
            data: {
                id: 123,
                name: 'Seattle SkyLine Poster',
                category: 'world',
                price: 20,
                stock: 10,
            },
            previousData: {
                id: 123,
                name: 'NewYork SkyLine Poster',
                category: 'cities',
                price: 20,
                stock: 10,
            },
        },
        author: {
            id: 789,
            fullName: 'Takeshi Kovacs',
            avatar: 'https://cdn.images.express.co.uk/img/dynamic/20/285x395/922644_1.jpg',
        },
    },
    {
        id: 4,
        resource: 'products',
        action: 'update',
        date: new Date(2020, 0, 6, 18, 48, 0),
        payload: {
            id: 789,
            data: {
                id: 123,
                name: 'San Francisco SkyLine Poster',
                category: 'cities',
                price: 50,
                stock: 10,
            },
            previousData: {
                id: 123,
                name: 'Seattle SkyLine Poster',
                category: 'wolrd',
                price: 20,
                stock: 10,
            },
        },
        author: {
            id: 789,
            fullName: 'Takeshi Kovacs',
            avatar: 'https://cdn.images.express.co.uk/img/dynamic/20/285x395/922644_1.jpg',
        },
    },
    {
        id: 5,
        resource: 'products',
        action: 'update',
        date: new Date(2020, 0, 6, 19, 48, 0),
        payload: {
            id: 789,
            data: {
                id: 123,
                name: 'San Francisco SkyLine Poster',
                category: 'cities',
                price: 50,
                stock: 10,
            },
            previousData: {
                id: 123,
                name: 'Seattle SkyLine Poster',
                category: 'wolrd',
                price: 20,
                stock: 9,
            },
        },
        author: {
            id: 789,
            fullName: 'Takeshi Kovacs',
            avatar: 'https://cdn.images.express.co.uk/img/dynamic/20/285x395/922644_1.jpg',
        },
    },
];
var i18nProvider = polyglotI18nProvider(function () {
    return mergeTranslations(englishMessages, raAuditLogLanguageEnglish);
});
var history = createMemoryHistory();
var store = function (dataProvider) {
    return createAdminStore({
        dataProvider: dataProvider,
        history: history,
        initialState: {
            admin: {
                resources: {
                    products: {},
                    events: {},
                },
            },
        },
    });
};
var dataProvider = function () {
    return Promise.resolve({
        data: records.sort(function (a, b) { return b.date.valueOf() - a.date.valueOf(); }),
        total: records.length,
    });
};
var neverLoadedDataProvider = convertLegacyDataProvider(function () {
    return new Promise(function () {
        /* */
    });
});
export var NotLoadedYet = function () { return (React.createElement(Provider, { store: store(neverLoadedDataProvider) },
    React.createElement(Router, { history: history },
        React.createElement(DataProviderContext.Provider, { value: neverLoadedDataProvider },
            React.createElement(Container, { maxWidth: "lg" },
                React.createElement(Box, { display: "flex", flexDirection: "column", margin: "auto", bgcolor: "background.default", padding: 2, maxWidth: "25%" },
                    React.createElement(RecordTimeline, null))))))); };
export var Loaded = function () { return (React.createElement(Provider, { store: store(dataProvider) },
    React.createElement(Router, { history: history },
        React.createElement(DataProviderContext.Provider, { value: convertLegacyDataProvider(dataProvider) },
            React.createElement(TranslationProvider, { locale: "en", i18nProvider: i18nProvider },
                React.createElement(Container, { maxWidth: "lg" },
                    React.createElement(Box, { display: "flex", flexDirection: "column", bgcolor: "background.default", padding: 2, margin: "auto", width: "25%" },
                        React.createElement(RecordTimeline, { record: { id: 123 } })))))))); };
var sortByDate = function (a, b) {
    return new Date(b.date).valueOf() - new Date(a.date).valueOf();
};
var getMonthForAuditLog = function (record) {
    var date = new Date(record.date);
    date.setMilliseconds(0);
    date.setSeconds(0);
    date.setMinutes(0);
    date.setHours(0);
    date.setDate(1);
    return date.toISOString();
};
var getMonthString = function (date, locale) {
    return new Date(date).toLocaleDateString(locale, {
        month: 'long',
        year: 'numeric',
    });
};
var groupByMonth = function (records, locale) {
    var initialMap = {};
    var map = records.reduce(function (acc, record) {
        var label = getMonthString(getMonthForAuditLog(record), locale);
        acc[label] = acc[label] || [];
        acc[label] = acc[label].concat(record);
        return acc;
    }, initialMap);
    return Object.keys(map).reduce(function (acc, label) {
        acc.push({ label: label, records: map[label].sort(sortByDate) });
        return acc;
    }, []);
};
export var CustomGrouping = function () { return (React.createElement(Provider, { store: store(dataProvider) },
    React.createElement(Router, { history: history },
        React.createElement(DataProviderContext.Provider, { value: convertLegacyDataProvider(dataProvider) },
            React.createElement(TranslationProvider, { locale: "en", i18nProvider: i18nProvider },
                React.createElement(Container, { maxWidth: "lg" },
                    React.createElement(Box, { display: "flex", flexDirection: "column", bgcolor: "background.default", padding: 2, margin: "auto", width: "25%" },
                        React.createElement(RecordTimeline, { record: { id: 123 }, groupLogs: groupByMonth })))))))); };
var useMyTimelineGroupStyles = makeStyles(function (theme) { return ({
    label: {
        color: theme.palette.primary.main,
        fontWeight: 'bold',
    },
    events: {
        borderRadius: 0,
        boxShadow: 'none',
    },
}); });
var MyTimelineGroup = function () {
    var classes = useMyTimelineGroupStyles();
    return React.createElement(TimelineGroup, { classes: classes });
};
export var CustomGroupComponent = function () { return (React.createElement(Provider, { store: store(dataProvider) },
    React.createElement(Router, { history: history },
        React.createElement(DataProviderContext.Provider, { value: convertLegacyDataProvider(dataProvider) },
            React.createElement(TranslationProvider, { locale: "en", i18nProvider: i18nProvider },
                React.createElement(Container, { maxWidth: "lg" },
                    React.createElement(Box, { display: "flex", flexDirection: "column", bgcolor: "background.default", padding: 2, width: "25%" },
                        React.createElement(RecordTimeline, { record: { id: 123 } },
                            React.createElement(MyTimelineGroup, null))))))))); };
var MyTimelineItem = function (props) {
    var record = useRecordContext(props.record);
    var eventLabel = useEventLabel({ record: record });
    return (React.createElement(ListItem, null,
        React.createElement(ListItemText, { primary: record.author.fullName, secondary: eventLabel })));
};
var MyTimelineGroupForCustomItem = function () {
    return (React.createElement(TimelineGroup, null,
        React.createElement(MyTimelineItem, null)));
};
export var CustomItemComponent = function () { return (React.createElement(Provider, { store: store(dataProvider) },
    React.createElement(Router, { history: history },
        React.createElement(DataProviderContext.Provider, { value: convertLegacyDataProvider(dataProvider) },
            React.createElement(TranslationProvider, { locale: "en", i18nProvider: i18nProvider },
                React.createElement(Container, { maxWidth: "lg" },
                    React.createElement(Box, { display: "flex", flexDirection: "column", bgcolor: "background.default", padding: 2, width: "25%" },
                        React.createElement(RecordTimeline, { record: { id: 123 } },
                            React.createElement(MyTimelineGroupForCustomItem, null))))))))); };
