// @react-admin/ra-rbac/lib/src/detail/ShowActions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShowActions = void 0;
var React = __importStar(require("react"));
var react_admin_1 = require("react-admin");
var usePermissions_1 = require("../usePermissions");
var canAccess_1 = require("../canAccess");
/**
 * Replacement for react-admin's ShowAction that adds RBAC control to actions
 *
 * Users must have the 'edit' permission on the resource and record to see the EditButton.
 *
 * @example
 * import { Show } from 'react-admin';
 * import { ShowActions } from '@react-admin/ra-rbac';
 *
 * export const PostShow = (props) => (
 *     <Show actions={<ShowActions />} {...props}>
 *         ...
 *     </Show>
 * );
 */
var ShowActions = function (props) {
    var className = props.className;
    var _a = react_admin_1.useShowContext(props), basePath = _a.basePath, record = _a.record, rest = __rest(_a, ["basePath", "record"]);
    var hasEdit = react_admin_1.useResourceDefinition(props).hasEdit;
    var resource = react_admin_1.useResourceContext(props);
    var _b = usePermissions_1.usePermissions(), loaded = _b.loaded, permissions = _b.permissions;
    return (loaded && (React.createElement(react_admin_1.TopToolbar, __assign({ className: className }, sanitizeRestProps(rest)), hasEdit &&
        canAccess_1.canAccess({
            permissions: permissions,
            action: 'edit',
            resource: resource,
            record: record,
        }) && React.createElement(react_admin_1.EditButton, { basePath: basePath, record: record }))));
};
exports.ShowActions = ShowActions;
exports.ShowActions.propTypes = react_admin_1.ShowActions.propTypes;
var sanitizeRestProps = function (_a) {
    var basePath = _a.basePath, defaultTitle = _a.defaultTitle, className = _a.className, hasCreate = _a.hasCreate, hasEdit = _a.hasEdit, hasList = _a.hasList, hasShow = _a.hasShow, loaded = _a.loaded, loading = _a.loading, refetch = _a.refetch, resource = _a.resource, rest = __rest(_a, ["basePath", "defaultTitle", "className", "hasCreate", "hasEdit", "hasList", "hasShow", "loaded", "loading", "refetch", "resource"]);
    return rest;
};
