// @react-admin/ra-enterprise/src/list/Filter.tsx
import * as React from 'react';
import { FC, ReactNode } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import {
    sanitizeListRestProps,
    useListContext,
    FilterButton,
    ClassesOverride,
} from 'react-admin';
import { FilterForm } from './FilterForm';

const useStyles = makeStyles(
    {
        button: {},
        form: {},
    },
    { name: 'RaFilter' }
);

export interface FilterProps {
    children: ReactNode;
    classes?: ClassesOverride<typeof useStyles>;
    context?: 'form' | 'button';
    variant?: string;
}

/**
 * A customized version of the Filter button/form combo from react-admin which
 * uses our custom FilterForm. This fixes the issue where the filter form is taking
 * too much vertical space even though there are no filters.
 *
 * @example
 *
 * const PostFilter = (props) => (
 *     <Filter {...props}>
 *         <TextInput label="Search" source="q" alwaysOn />
 *         <TextInput label="Title" source="title" defaultValue="Hello, World!" />
 *     </Filter>
 * );
 *
 * export const PostList = (props) => (
 *     <List {...props} filters={<PostFilter />}>
 *         ...
 *     </List>
 * );
 *
 */
export const Filter: FC<FilterProps> = props => {
    const classes = useStyles(props);
    const {
        resource,
        showFilter,
        hideFilter,
        setFilters,
        displayedFilters,
        filterValues,
    } = useListContext(props);
    const renderButton = () => {
        const {
            classes: classesOverride,
            context,
            children,
            variant,
            ...rest
        } = props;

        return (
            <FilterButton
                className={classes.button}
                resource={resource}
                filters={React.Children.toArray(children)}
                showFilter={showFilter}
                displayedFilters={displayedFilters}
                filterValues={filterValues}
                {...sanitizeListRestProps(rest)}
            />
        );
    };

    const renderForm = () => {
        const { classes: classesOverride, context, children, ...rest } = props;

        return (
            <FilterForm
                className={classes.form}
                resource={resource}
                filters={React.Children.toArray(children)}
                hideFilter={hideFilter}
                displayedFilters={displayedFilters}
                initialValues={filterValues}
                setFilters={setFilters}
                {...sanitizeListRestProps(rest)}
            />
        );
    };

    return props.context === 'button' ? renderButton() : renderForm();
};

Filter.propTypes = {
    children: PropTypes.node,
    classes: PropTypes.object,
    context: PropTypes.oneOf(['form', 'button']),
};
