// @react-admin/ra-enterprise/src/list/ListToolbar.tsx
import * as React from 'react';
import { ReactElement } from 'react';
import PropTypes from 'prop-types';
import { Toolbar } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import {
    FilterContext,
    ListToolbar as RaListToolbar,
    ListToolbarProps as RaListToolbarProps,
} from 'react-admin';
import { BreadcrumbProps } from '@react-admin/ra-navigation';
import { FilterForm } from './FilterForm';

const ListToolbar = (props: ListToolbarProps) => {
    const { classes: classesOverride, filters, actions, ...rest } = props;
    const classes = useStyles(props);

    return Array.isArray(filters) ? (
        <FilterContext.Provider value={filters}>
            <Toolbar className={classes.toolbar}>
                <span />
                {actions &&
                    React.cloneElement(actions, {
                        ...rest,
                        className: classes.actions,
                        ...actions.props,
                    })}
            </Toolbar>
            <FilterForm />
        </FilterContext.Provider>
    ) : (
        <>
            <Toolbar className={classes.toolbar}>
                {actions &&
                    React.cloneElement(actions, {
                        ...rest,
                        className: classes.actions,
                        filters,
                        ...actions.props,
                    })}
            </Toolbar>
            {filters &&
                React.cloneElement(filters, {
                    ...rest,
                    context: 'form',
                })}
        </>
    );
};

ListToolbar.propTypes = {
    // @ts-ignore
    ...RaListToolbar.propTypes,
    breadcrumb: PropTypes.element,
};

export interface ListToolbarProps extends RaListToolbarProps {
    breadcrumb?: ReactElement<BreadcrumbProps>;
}

const useStyles = makeStyles(
    theme => ({
        toolbar: {
            justifyContent: 'space-between',
            alignItems: 'flex-start',
            paddingRight: 0,
            [theme.breakpoints.up('xs')]: {
                paddingLeft: 0,
            },
            [theme.breakpoints.down('xs')]: {
                paddingLeft: theme.spacing(2),
                backgroundColor: theme.palette.background.paper,
            },
        },
        actions: {
            alignItems: 'center',
            paddingTop: theme.spacing(3),
            minHeight: theme.spacing(5),
            [theme.breakpoints.down('xs')]: {
                padding: theme.spacing(1),
                backgroundColor: theme.palette.background.paper,
            },
            width: '100%',
        },
    }),
    { name: 'RaListToolbar' }
);

export default React.memo(ListToolbar);
