// @react-admin/ra-navigation/esm/stories/MenuItemCategory.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import * as React from 'react';
import { useEffect } from 'react';
import { Admin, Resource, Layout, List, Datagrid, TextField, DateField, } from 'react-admin';
import { createMemoryHistory } from 'history';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import DashboardIcon from '@material-ui/icons/Dashboard';
import MusicIcon from '@material-ui/icons/MusicNote';
import PeopleIcon from '@material-ui/icons/People';
import SettingsIcon from '@material-ui/icons/Settings';
import FlashIcon from '@material-ui/icons/FlashOn';
import { makeStyles } from '@material-ui/core/styles';
import merge from 'lodash/merge';
import { AppLocationContext, useAppLocationState, useDefineAppLocation, useResourceAppLocation, } from '../src/app-location';
import { Menu, MenuItem, MenuItemCategory, MultiLevelMenu, theme, } from '../src/multi-level-menu';
import dataProvider from './dataProvider';
export default { title: 'ra-navigation/MultiLevelMenu/MenuItemCategory' };
var Dashboard = function () {
    useDefineAppLocation('dashboard');
    return (React.createElement(Card, null,
        React.createElement("h1", null, "Dashboard")));
};
var SongList = function (props) { return (React.createElement(List, __assign({}, props),
    React.createElement(Datagrid, null,
        React.createElement(TextField, { source: "id" }),
        React.createElement(TextField, { source: "title" }),
        React.createElement(DateField, { source: "released" })))); };
var ArtistList = function (props) { return (React.createElement(List, __assign({}, props),
    React.createElement(ArtistsDatagrid, null))); };
var types = {
    Rock: 'artists.rock',
    'Folk Rock': 'artists.rock.folk',
    'Pop Rock': 'artists.rock.pop',
    Jazz: 'artists.jazz',
    RB: 'artists.jazz.rb',
};
var ArtistsDatagrid = function (props) {
    var _a = useAppLocationState(), setLocation = _a[1];
    var resourceLocation = useResourceAppLocation();
    useEffect(function () {
        var type = props.filterValues.type;
        if (typeof type !== 'undefined') {
            setLocation(types[type]);
        }
    }, 
    /* eslint-disable react-hooks/exhaustive-deps */
    [
        JSON.stringify({
            resourceLocation: resourceLocation,
            filter: props.filterValues,
        }),
    ]
    /* eslint-enable react-hooks/exhaustive-deps */
    );
    return (React.createElement(Datagrid, __assign({}, props),
        React.createElement(TextField, { source: "id" }),
        React.createElement(TextField, { source: "name" })));
};
var useStyles = makeStyles({
    configuration: {
        marginTop: 'auto',
    },
});
var MultiLevelMenuWithCategories = function () {
    var classes = useStyles();
    return (React.createElement(MultiLevelMenu, { variant: "categories" },
        React.createElement(MenuItemCategory, { name: "dashboard", to: "/", exact: true, label: "Dashboard", icon: React.createElement(DashboardIcon, null) }),
        React.createElement(MenuItemCategory, { name: "songs", icon: React.createElement(MusicIcon, null), to: "/songs", label: "Songs" }),
        React.createElement(MenuItemCategory, { name: "artists", label: "Artists", icon: React.createElement(PeopleIcon, null) },
            React.createElement(CardContent, null,
                React.createElement(Typography, { variant: "h4", gutterBottom: true }, "All artists"),
                React.createElement(Menu, null,
                    React.createElement(MenuItem, { name: "artists", to: '/artists?filter={}', label: "All Artists" }),
                    React.createElement(MenuItem, { name: "artists.rock", to: '/artists?filter={"type":"Rock"}', label: "Rock" },
                        React.createElement(MenuItem, { name: "artists.rock.pop", to: '/artists?filter={"type":"Pop Rock"}', label: "Pop Rock" }),
                        React.createElement(MenuItem, { name: "artists.rock.folk", to: '/artists?filter={"type":"Folk Rock"}', label: "Folk Rock" })),
                    React.createElement(MenuItem, { name: "artists.jazz", to: '/artists?filter={"type":"Jazz"}', label: "Jazz" },
                        React.createElement(MenuItem, { name: "artists.jazz.rb", to: '/artists?filter={"type":"RB"}', label: "R&B" }))))),
        React.createElement(MenuItemCategory, { name: "rock_artists", label: "Rock Artists", icon: React.createElement(FlashIcon, null) },
            React.createElement(CardContent, null,
                React.createElement(Typography, { variant: "h4", gutterBottom: true }, "Rock artists"),
                React.createElement(Menu, null,
                    React.createElement(MenuItem, { name: "rock_artists.pop", to: '/artists?filter={"type":"Pop Rock"}', label: "Pop Rock" }),
                    React.createElement(MenuItem, { name: "rock_artists.folk", to: '/artists?filter={"type":"Folk Rock"}', label: "Folk Rock" })))),
        React.createElement(MenuItemCategory, { className: classes.configuration, name: "configuration", to: "/", exact: true, label: "Configuration", icon: React.createElement(SettingsIcon, null) })));
};
var LayoutWithCategories = function (props) {
    return (React.createElement(AppLocationContext, null,
        React.createElement(Layout, __assign({}, props, { menu: MultiLevelMenuWithCategories }))));
};
export var Basic = function () { return (React.createElement(Admin, { history: createMemoryHistory(), dataProvider: dataProvider, layout: LayoutWithCategories, dashboard: Dashboard, theme: theme },
    React.createElement(Resource, { name: "songs", list: SongList }),
    React.createElement(Resource, { name: "artists", list: ArtistList }))); };
export var DarkMode = function () {
    var darkTheme = merge({}, theme, {
        palette: {
            type: 'dark',
        },
    });
    return (React.createElement(Admin, { history: createMemoryHistory(), dataProvider: dataProvider, layout: LayoutWithCategories, dashboard: Dashboard, theme: darkTheme },
        React.createElement(Resource, { name: "songs", list: SongList }),
        React.createElement(Resource, { name: "artists", list: ArtistList })));
};
var LongMultiLevelMenuWithCategories = function () { return (React.createElement(MultiLevelMenu, { variant: "categories" },
    React.createElement(MenuItemCategory, { name: "dashboard", to: "/", exact: true, label: "Dashboard", icon: React.createElement(DashboardIcon, null) }),
    React.createElement(MenuItemCategory, { name: "songs", icon: React.createElement(MusicIcon, null), to: "/songs", label: "Songs" }),
    React.createElement(MenuItemCategory, { name: "artists", label: "Artists", icon: React.createElement(PeopleIcon, null) }),
    React.createElement(MenuItemCategory, { name: "rock_artists", label: "Rock Artists", icon: React.createElement(FlashIcon, null) }),
    React.createElement(MenuItemCategory, { name: "foo1", label: "Foo1", icon: React.createElement(MusicIcon, null), to: "/foo1" }),
    React.createElement(MenuItemCategory, { name: "foo2", label: "Foo2", icon: React.createElement(MusicIcon, null), to: "/foo2" }),
    React.createElement(MenuItemCategory, { name: "foo3", label: "Foo3", icon: React.createElement(MusicIcon, null), to: "/foo3" }),
    React.createElement(MenuItemCategory, { name: "foo4", label: "Foo4", icon: React.createElement(MusicIcon, null), to: "/foo4" }),
    React.createElement(MenuItemCategory, { name: "foo5", label: "Foo5", icon: React.createElement(MusicIcon, null), to: "/foo5" }),
    React.createElement(MenuItemCategory, { name: "foo6", label: "Foo6", icon: React.createElement(MusicIcon, null), to: "/foo6" }),
    React.createElement(MenuItemCategory, { name: "foo7", label: "Foo7", icon: React.createElement(MusicIcon, null), to: "/foo7" }),
    React.createElement(MenuItemCategory, { name: "foo8", label: "Foo8", icon: React.createElement(MusicIcon, null), to: "/foo8" }),
    React.createElement(MenuItemCategory, { name: "foo9", label: "Foo9", icon: React.createElement(MusicIcon, null), to: "/foo9" }),
    React.createElement(MenuItemCategory, { name: "foo10", label: "Foo10", icon: React.createElement(MusicIcon, null), to: "/foo10" }),
    React.createElement(MenuItemCategory, { name: "foo11", label: "Foo11", icon: React.createElement(MusicIcon, null), to: "/foo11" }),
    React.createElement(MenuItemCategory, { name: "foo12", label: "Foo12", icon: React.createElement(MusicIcon, null), to: "/foo12" }),
    React.createElement(MenuItemCategory, { name: "foo13", label: "Foo13", icon: React.createElement(MusicIcon, null), to: "/foo13" }),
    React.createElement(MenuItemCategory, { name: "foo14", label: "Foo14", icon: React.createElement(MusicIcon, null), to: "/foo14" }),
    React.createElement(MenuItemCategory, { name: "foo15", label: "Foo15", icon: React.createElement(MusicIcon, null), to: "/foo15" }),
    React.createElement(MenuItemCategory, { name: "foo16", label: "Foo16", icon: React.createElement(MusicIcon, null), to: "/foo16" }),
    React.createElement(MenuItemCategory, { name: "foo17", label: "Foo17", icon: React.createElement(MusicIcon, null), to: "/foo17" }),
    React.createElement(MenuItemCategory, { name: "foo18", label: "Foo18", icon: React.createElement(MusicIcon, null), to: "/foo18" }),
    React.createElement(MenuItemCategory, { name: "foo19", label: "Foo19", icon: React.createElement(MusicIcon, null), to: "/foo19" }),
    React.createElement(MenuItemCategory, { name: "foo10", label: "Foo10", icon: React.createElement(MusicIcon, null), to: "/foo10" }))); };
var LayoutWithLongMenuWithCategories = function (props) {
    return (React.createElement(AppLocationContext, null,
        React.createElement(Layout, __assign({}, props, { menu: LongMultiLevelMenuWithCategories }))));
};
export var Long = function () {
    var darkTheme = merge({}, theme, {
        palette: {
            type: 'dark',
        },
    });
    return (React.createElement(Admin, { history: createMemoryHistory(), dataProvider: dataProvider, layout: LayoutWithLongMenuWithCategories, dashboard: Dashboard, theme: darkTheme },
        React.createElement(Resource, { name: "songs", list: SongList }),
        React.createElement(Resource, { name: "artists", list: ArtistList })));
};
