// @react-admin/ra-realtime/src/Warning.tsx
import React from 'react';
import { useTranslate, RefreshButton } from 'react-admin';
import { Paper, Typography } from '@material-ui/core';
import { darken, lighten, makeStyles } from '@material-ui/core/styles';

export const Warning = (props: WarningProps) => {
    const { message, refresh, onRefresh } = props;
    const classes = useStyles(props);
    const translate = useTranslate();

    return (
        <Paper className={classes.root} role="alert" square elevation={0}>
            <div className={classes.message}>
                <Typography gutterBottom component="div">
                    {translate('ra-realtime.notification.title')}
                </Typography>
                {message}
            </div>
            <div className={classes.action}>
                {refresh && <RefreshButton onClick={onRefresh} />}
            </div>
        </Paper>
    );
};

export type WarningProps = {
    message: string;
    refresh?: boolean;
    onRefresh?: () => void;
};

const useStyles = makeStyles(
    theme => {
        const getColor = theme.palette.type === 'light' ? darken : lighten;
        const getBackgroundColor =
            theme.palette.type === 'light' ? lighten : darken;
        return {
            root: {
                ...theme.typography.body2,
                borderRadius: theme.shape.borderRadius,
                display: 'flex',
                padding: '6px 16px',
                margin: `${theme.spacing(1)}px 0`,
                color: getColor(theme.palette.warning.main, 0.6),
                backgroundColor: getBackgroundColor(
                    theme.palette.warning.main,
                    0.9
                ),
            },
            message: {
                padding: `${theme.spacing(1)}px 0`,
            },
            action: {
                display: 'flex',
                alignItems: 'center',
                marginLeft: 'auto',
                paddingLeft: theme.spacing(2),
                marginRight: -theme.spacing(1),
            },
        };
    },
    {
        name: 'RaWarning',
    }
);
