// @react-admin/ra-editable-datagrid/src/EditableDatagridRow.tsx
import * as React from 'react';
import {
    cloneElement,
    MouseEvent,
    ReactElement,
    ReactNode,
    useCallback,
    useMemo,
    useState,
} from 'react';
import {
    CRUD_UPDATE,
    DatagridRow,
    Identifier,
    Record,
    SaveContextProvider,
    useNotify,
    useDatagridStyles,
    useSaveModifiers,
    useUpdate,
    useResourceContext,
    useRecordContext,
} from 'react-admin';

import { makeStyles } from '@material-ui/core';

import { EditRowContext } from './EditRowContext';

const EditableRow = (props: EditableRowProps) => {
    const {
        form,
        resource: resourceOverride,
        id,
        record: recordOverride,
        basePath,
        children,
        rowClick,
        mutationMode,
        undoable,
        ...rest
    } = props;
    const resource = useResourceContext(props);
    const record = useRecordContext(props);
    const [isEdit, setEdit] = useState(false);
    const classes = useStyles(emptyObject);
    const openEditMode = useCallback((): void => {
        setEdit(true);
    }, []);
    const closeEditMode = useCallback((): void => {
        setEdit(false);
    }, []);
    const notify = useNotify();
    const [update, { loading: saving }] = useUpdate(
        resource,
        id,
        {}, // set by the caller
        record
    );

    const defaultOnSuccess = useCallback(() => {
        notify(
            'ra.notification.updated',
            'info',
            {
                smart_count: 1,
            },
            undoable
        );
        closeEditMode();
    }, [notify, undoable, closeEditMode]);

    const defaultOnFailure = useCallback(
        (error): void => {
            notify(
                typeof error === 'string'
                    ? error
                    : error.message || 'ra.notification.http_error',
                'warning'
            );
        },
        [notify]
    );

    const {
        onSuccessRef,
        setOnSuccess,
        onFailureRef,
        setOnFailure,
        transformRef,
        setTransform,
    } = useSaveModifiers({
        onFailure: defaultOnFailure,
        onSuccess: defaultOnSuccess,
    });

    const handleClick = useCallback(
        (event: MouseEvent<HTMLElement>): void => {
            const { tbody, row, column } = getTableClickEventPosition(event);
            openEditMode();
            // once the row is replaced by a form, focus the input inside the cell clicked
            setTimeout(() => {
                // No way to know the markup of the form in advance, as developers
                // can inject a form element of their own. The only valid assumption
                // is that the form should have the same number of columns as the row.
                // So we select the input based on the column it's in.
                const input = tbody.querySelector(
                    `tr:nth-child(${row}) td:nth-child(${column}) input`
                ) as HTMLInputElement;
                input && input.focus && input.focus();
            }, 100); // FIXME not super robust
        },
        [openEditMode]
    );

    const save = useCallback(
        (
            data: Partial<Record>,
            _, // unused redirectTo
            {
                onSuccess: onSuccessFromSave,
                onFailure: onFailureFromSave,
                transform: transformFromSave,
            } = {}
        ) =>
            Promise.resolve(
                transformFromSave
                    ? transformFromSave(data)
                    : !!transformRef.current
                    ? transformRef.current(data)
                    : data
            ).then(data =>
                update(
                    { payload: { data } },
                    {
                        action: CRUD_UPDATE,
                        onSuccess: onSuccessFromSave
                            ? onSuccessFromSave
                            : !!onSuccessRef.current
                            ? onSuccessRef.current
                            : defaultOnSuccess,
                        onFailure: onFailureFromSave
                            ? onFailureFromSave
                            : !!onFailureRef.current
                            ? onFailureRef.current
                            : defaultOnFailure,
                        undoable,
                    }
                )
            ),
        [
            defaultOnFailure,
            defaultOnSuccess,
            update,
            undoable,
            onFailureRef,
            onSuccessRef,
            transformRef,
        ]
    );

    const saveContext = {
        save,
        saving,
        setOnSuccess,
        setOnFailure,
        setTransform,
        onSuccessRef,
        onFailureRef,
        transformRef,
    };

    const rowContext = useMemo(
        () => ({
            open: openEditMode,
            close: closeEditMode,
        }),
        [openEditMode, closeEditMode]
    );

    return (
        <EditRowContext.Provider value={rowContext}>
            {isEdit ? (
                <SaveContextProvider value={saveContext}>
                    {cloneElement(form, {
                        id,
                        quitEditMode: closeEditMode,
                        resource,
                        record,
                        basePath,
                        save,
                        saving,
                        mutationMode,
                        undoable,
                        ...rest,
                    })}
                </SaveContextProvider>
            ) : (
                <DatagridRow
                    resource={resource}
                    id={id}
                    record={record}
                    basePath={basePath}
                    {...rest}
                    className={classes.td}
                    onClick={
                        rowClick === 'edit'
                            ? handleClick
                            : (): void => undefined
                    }
                >
                    {children}
                </DatagridRow>
            )}
        </EditRowContext.Provider>
    );
};

/**
 * Based on a MouseEvent triggered by a click on a table row,
 * get the tbody element, the row and column number of the cell clicked.
 *
 * @param {MouseEvent} event
 */
const getTableClickEventPosition = (
    event: MouseEvent<HTMLElement>
): { tbody: HTMLElement; row: number; column: number } => {
    const target = event.target as HTMLElement;
    const td = target.closest('td');
    const tr = td.parentNode;
    const columns = tr.children as HTMLCollection;
    let column: number;
    for (let index = 0; index < columns.length; index++) {
        if (columns.item(index) === td) {
            column = index + 1;
        }
    }
    const tbody = tr.parentNode as HTMLElement;
    const rows = tbody.children as HTMLCollection;
    let row: number;
    for (let index = 0; index < rows.length; index++) {
        if (rows.item(index) === tr) {
            row = index + 1;
        }
    }
    return { tbody, row, column };
};

export interface EditableRowProps {
    classes?: ReturnType<typeof useDatagridStyles>;
    form: ReactElement;
    resource?: string;
    id?: Identifier;
    record?: Record;
    basePath?: string;
    children?: ReactNode;
    undoable?: boolean;
    rowClick?: string;
    [key: string]: any;
}

const useStyles = makeStyles(
    {
        td: {
            '& td:last-of-type > *': {
                visibility: 'hidden',
            },
            '&:hover td:last-of-type > *': {
                visibility: 'visible',
            },
        },
    },
    {
        name: 'RaEditableDatagridRow',
    }
);

const emptyObject = {};

export default EditableRow;
