// @react-admin/ra-markdown/src/MarkdownInput.tsx
import React, { useEffect, useRef, RefObject } from 'react';
import merge from 'lodash/merge';
import { InputHelperText, InputProps, Labeled, useInput } from 'react-admin';
import { FormHelperText } from '@material-ui/core';

import 'codemirror/lib/codemirror.css';
import '@toast-ui/editor/dist/toastui-editor.css';
import { Editor } from '@toast-ui/react-editor';

export interface MarkdownInputProps
    extends Partial<
            Omit<
                React.ComponentProps<typeof Editor>,
                'onChange' | 'onBlur' | 'onFocus'
            >
        >,
        InputProps {
    helperText?: string;
    label?: string;
    fullWidth?: boolean;
}

/**
 * An Input component for markdown content
 *
 * @example
 *
 *    import { Edit, SimpleForm, TextInput } from 'react-admin';
 *    import { MarkdownInput } from '@react-admin/ra-markdown';
 *
 *    const PostEdit = props => (
 *        <Edit {...props}>
 *            <SimpleForm>
 *                <TextInput source="title" />
 *                <MarkdownInput source="description" />
 *            </SimpleForm>
 *        </Edit>
 *    );
 *
 * Based on the '@toast-ui/react-editor' package.
 *
 * @example
 *
 *    const options = {
 *        previewStyle: 'tab',
 *        height: 'auto',
 *        initialEditType: 'markdown',
 *        hideModeSwitch: true,
 *    };
 *    <MarkdownInput source="description" options={options} />
 *
 * The object passed as `options` props is passed to the <Editor> component
 * @see https://nhn.github.io/tui.editor/latest/ToastUIEditor
 */
export const MarkdownInput = ({
    fullWidth,
    helperText,
    label,
    options,
    resource,
    source,
    ...rest
}: MarkdownInputProps) => {
    const editorRef: RefObject<Editor> = useRef();

    const {
        input: { onChange, value },
        isRequired,
        meta,
    } = useInput({
        source,
        ...rest,
    });

    const { error, submitError, touched } = meta;

    const handleChange = editor => {
        onChange(editor.getMarkdown());
    };

    useEffect(() => {
        // We have to use the addHook function of the Editor to catch changes.
        // https://nhn.github.io/tui.editor/latest/ToastUIEditor#addHook
        // This effect will be triggered only one time when the Editor is rendered.
        if (editorRef && editorRef.current) {
            const instance = editorRef.current.getInstance();
            instance.addHook('change', () => handleChange(instance));
        }
    }, [editorRef]); // eslint-disable-line react-hooks/exhaustive-deps

    const mergedOptions = merge({}, defaultOptions, options);

    return (
        <Labeled
            fullWidth={fullWidth}
            isRequired={isRequired}
            label={label}
            meta={meta}
            resource={resource}
            source={source}
        >
            <>
                <Editor
                    initialValue={value}
                    ref={editorRef}
                    {...mergedOptions}
                />
                <FormHelperText error={touched && (error || submitError)}>
                    <InputHelperText
                        touched={touched}
                        error={error || submitError}
                        helperText={helperText}
                    />
                </FormHelperText>
            </>
        </Labeled>
    );
};

const defaultOptions = {
    previewStyle: 'vertical',
    height: '512px',
    initialEditType: 'wysiwyg',
    useCommandShortcut: true,
};

MarkdownInput.defaultProps = {
    fullWidth: true,
    options: defaultOptions,
};
