// @react-admin/ra-enterprise/lib/src/list/ListActions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListActions = void 0;
var React = __importStar(require("react"));
var react_1 = require("react");
var prop_types_1 = __importDefault(require("prop-types"));
var react_admin_1 = require("react-admin");
var layout_1 = require("../layout");
var ListActions = function (props) {
    var breadcrumb = props.breadcrumb, className = props.className, exporter = props.exporter, filtersProp = props.filters, rest = __rest(props, ["breadcrumb", "className", "exporter", "filters"]);
    var _a = react_admin_1.useListContext(props), currentSort = _a.currentSort, displayedFilters = _a.displayedFilters, filterValues = _a.filterValues, basePath = _a.basePath, selectedIds = _a.selectedIds, showFilter = _a.showFilter, total = _a.total;
    var resource = react_admin_1.useResourceContext(rest);
    var filters = react_1.useContext(react_admin_1.FilterContext) || filtersProp;
    var hasCreate = react_admin_1.useResourceDefinition(rest).hasCreate;
    return react_1.useMemo(function () { return (React.createElement(react_admin_1.TopToolbar, __assign({ className: className }, react_admin_1.sanitizeListRestProps(rest)),
        breadcrumb,
        filtersProp
            ? react_1.cloneElement(filtersProp, {
                resource: resource,
                showFilter: showFilter,
                displayedFilters: displayedFilters,
                filterValues: filterValues,
                context: 'button',
            })
            : filters && React.createElement(react_admin_1.FilterButton, null),
        hasCreate && React.createElement(react_admin_1.CreateButton, { basePath: basePath }),
        exporter !== false && (React.createElement(react_admin_1.ExportButton, { disabled: total === 0, resource: resource, sort: currentSort, filterValues: filterValues })))); }, [resource, displayedFilters, filterValues, selectedIds, filters, total] // eslint-disable-line react-hooks/exhaustive-deps
    );
};
exports.ListActions = ListActions;
exports.ListActions.propTypes = __assign(__assign({}, react_admin_1.ListActions.propTypes), { breadcrumb: prop_types_1.default.element });
exports.ListActions.defaultProps = {
    breadcrumb: React.createElement(layout_1.Breadcrumb, { variant: "actions" }),
    selectedIds: [],
    onUnselectItems: function () { return null; },
};
