// @react-admin/ra-calendar/esm/stories/dataProvider.js
import fakeRestProvider from 'ra-data-fakerest';
import { add, sub, set } from 'date-fns';
var today = set(new Date(), { minutes: 0, seconds: 0, milliseconds: 0 });
var dataProvider = fakeRestProvider({
    events: [
        {
            id: '0',
            title: 'Mail package',
            start: sub(set(today, { hours: 14 }), {
                days: 21,
            }).toISOString(),
            end: sub(set(today, { hours: 14, minutes: 45 }), {
                days: 21,
            }).toISOString(),
        },
        {
            id: '1',
            title: 'Initialize contract',
            start: sub(set(today, { hours: 9 }), {
                days: 13,
            }).toISOString(),
            end: sub(set(today, { hours: 10, minutes: 45 }), {
                days: 13,
            }).toISOString(),
        },
        {
            id: '2',
            title: 'Data training',
            start: sub(set(today, { hours: 10 }), {
                days: 8,
            }).toISOString(),
            end: sub(set(today, { hours: 11, minutes: 30 }), {
                days: 8,
            }).toISOString(),
        },
        {
            id: '3',
            title: 'Contract negociation',
            start: sub(set(today, { hours: 9 }), {
                days: 5,
            }).toISOString(),
            end: sub(set(today, { hours: 17, minutes: 30 }), {
                days: 5,
            }).toISOString(),
            backgroundColor: 'orange',
            borderColor: 'orange',
        },
        {
            id: '4',
            title: 'Package delivery',
            start: sub(set(today, { hours: 11 }), {
                days: 3,
            }).toISOString(),
            end: sub(set(today, { hours: 11, minutes: 30 }), {
                days: 3,
            }).toISOString(),
        },
        {
            id: '5',
            title: 'Interview Simon',
            start: sub(set(today, { hours: 15 }), {
                days: 2,
            }).toISOString(),
            end: sub(set(today, { hours: 17, minutes: 30 }), {
                days: 2,
            }).toISOString(),
        },
        {
            id: '6',
            title: 'Conf call John',
            start: sub(set(today, { hours: 9 }), {
                days: 1,
            }).toISOString(),
            end: sub(set(today, { hours: 12 }), {
                days: 1,
            }).toISOString(),
        },
        {
            id: '7',
            title: 'Meet Jane',
            start: sub(set(today, { hours: 13 }), {
                days: 1,
            }).toISOString(),
            end: sub(set(today, { hours: 15 }), {
                days: 1,
            }).toISOString(),
        },
        {
            id: '8',
            title: 'Sign contracts',
            start: set(today, { hours: 10 }).toISOString(),
            end: set(today, { hours: 11, minutes: 30 }).toISOString(),
        },
        {
            id: '9',
            title: 'Onboard John',
            start: set(today, { hours: 11 }).toISOString(),
            end: set(today, { hours: 12 }).toISOString(),
        },
        {
            id: '10',
            title: 'Open forum',
            start: set(today, { hours: 0 }).toISOString(),
            allDay: true,
        },
        {
            id: '11',
            title: '1&1 Carol',
            start: add(set(today, { hours: 11, minutes: 15 }), {
                days: 1,
            }).toISOString(),
            end: add(set(today, { hours: 11, minutes: 45 }), {
                days: 1,
            }).toISOString(),
        },
        {
            id: '12',
            title: 'Call Peter',
            start: add(set(today, { hours: 14, minutes: 30 }), {
                days: 1,
            }).toISOString(),
            end: add(set(today, { hours: 15 }), {
                days: 1,
            }).toISOString(),
        },
        {
            id: '13',
            title: 'Prepare interview Mickael',
            start: add(set(today, { hours: 10, minutes: 30 }), {
                days: 2,
            }).toISOString(),
            end: add(set(today, { hours: 12 }), {
                days: 2,
            }).toISOString(),
        },
        {
            id: '14',
            title: 'Work session project G',
            start: add(set(today, { hours: 9, minutes: 30 }), {
                days: 3,
            }).toISOString(),
            end: add(set(today, { hours: 16 }), {
                days: 3,
            }).toISOString(),
        },
        {
            id: '15',
            title: 'Call Janet',
            start: add(set(today, { hours: 16 }), {
                days: 5,
            }).toISOString(),
            end: add(set(today, { hours: 16, minutes: 45 }), {
                days: 5,
            }).toISOString(),
        },
        {
            id: '16',
            title: 'Demo day',
            start: add(set(today, { hours: 9 }), {
                days: 8,
            }).toISOString(),
            end: add(set(today, { hours: 17, minutes: 30 }), {
                days: 8,
            }).toISOString(),
        },
        {
            id: '17',
            title: 'Banana Conference',
            start: add(set(today, { hours: 9 }), {
                days: 13,
            }).toISOString(),
            end: add(set(today, { hours: 18 }), {
                days: 14,
            }).toISOString(),
        },
        {
            id: '18',
            title: 'Wine tasting',
            start: add(set(today, { hours: 18 }), {
                days: 21,
            }).toISOString(),
            end: add(set(today, { hours: 20, minutes: 30 }), {
                days: 21,
            }).toISOString(),
        },
        {
            id: '19',
            title: 'Onboard Tom',
            start: set(today, { hours: 12 }).toISOString(),
            end: set(today, { hours: 13 }).toISOString(),
        },
        {
            id: '20',
            title: 'Onboard Alex',
            start: set(today, { hours: 13 }).toISOString(),
            end: set(today, { hours: 14 }).toISOString(),
        },
        {
            id: '21',
            title: 'Onboard Marie',
            start: set(today, { hours: 15 }).toISOString(),
            end: set(today, { hours: 16 }).toISOString(),
        },
        {
            id: '22',
            title: 'Onboard Bob',
            start: set(today, { hours: 16 }).toISOString(),
            end: set(today, { hours: 17 }).toISOString(),
        },
        {
            id: '23',
            title: 'Onboard Rachel',
            start: set(today, { hours: 18 }).toISOString(),
            end: set(today, { hours: 19 }).toISOString(),
        },
    ],
}, true);
export default dataProvider;
