// @react-admin/ra-rbac/esm/stories/form/SimpleForm.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React from 'react';
import { Admin, Resource, ListGuesser, Edit, TextInput } from 'react-admin';
import fakeRestDataProvider from 'ra-data-fakerest';
import generateData from 'data-generator-retail';
import { SimpleForm } from '../../src';
export default { title: 'ra-rbac/form/SimpleForm' };
var data = generateData();
var dataProvider = fakeRestDataProvider(data, true);
var authProvider = {
    checkAuth: function () { return Promise.resolve(); },
    login: function () { return Promise.resolve(); },
    logout: function () { return Promise.resolve(); },
    checkError: function () { return Promise.resolve(); },
    getPermissions: function () {
        return promiseFor({
            permissions: [
                // 'delete' is missing
                { action: ['list', 'edit'], resource: 'products' },
                { action: 'write', resource: 'products.reference' },
                { action: 'write', resource: 'products.width' },
                { action: 'write', resource: 'products.height' },
                // 'products.description' is missing
                { action: 'write', resource: 'products.thumbnail' },
                // 'products.image' is missing
            ],
        });
    },
};
var ProductEdit = function (props) { return (React.createElement(Edit, __assign({}, props),
    React.createElement(SimpleForm, null,
        React.createElement(TextInput, { source: "reference" }),
        React.createElement(TextInput, { source: "width" }),
        React.createElement(TextInput, { source: "height" }),
        React.createElement(TextInput, { source: "description" }),
        React.createElement(TextInput, { source: "image" }),
        React.createElement(TextInput, { source: "thumbnail" })))); };
export var Basic = function () { return (React.createElement(Admin, { authProvider: authProvider, dataProvider: dataProvider },
    React.createElement(Resource, { name: "products", list: ListGuesser, edit: ProductEdit }),
    React.createElement(Resource, { name: "categories" }))); };
var promiseFor = function (result) {
    return new Promise(function (resolve) {
        setTimeout(function () {
            resolve(result);
        }, 300);
    });
};
