// @react-admin/ra-audit-log/src/EventAvatar.tsx
import * as React from 'react';
import { ReactElement } from 'react';
import { Avatar, AvatarProps } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import PersonIcon from '@material-ui/icons/Person';
import classnames from 'classnames';

/**
 * A Material-UI Avatar with fallback to the Person icon if no url was provided for an image.
 */
export const EventAvatar = (props: EventAvatarProps): ReactElement => {
    const { className, src, fullName, size = 'medium', ...rest } = props;
    const classes = useStyles(props);

    return (
        <Avatar
            src={src}
            className={classnames(className, classes[size])}
            {...rest}
        >
            {!!src ? null : fullName ? getInitials(fullName) : <PersonIcon />}
        </Avatar>
    );
};

const useStyles = makeStyles({
    small: {
        width: 20,
        height: 20,
    },
    medium: {
        width: 40,
        height: 40,
    },
});

export interface EventAvatarProps extends AvatarProps {
    fullName?: string;
    size?: 'small' | 'medium';
}

export const getInitials = (name: string): string =>
    name
        .match(/(^\w\w?|\b\w)?/g)
        .join('')
        .match(/(^\w|\w$)?/g)
        .join('')
        .toUpperCase();
