// @react-admin/ra-audit-log/src/useEventLabel.ts
import {
    useRecordContext,
    useTranslate,
    getFieldLabelTranslationArgs,
} from 'react-admin';
import { EventRecord } from './types';
import { useGetResourceLabel } from './useGetResourceLabel';

/**
 * Returns a translated label for an event.
 * @param options The hook options
 * @param options.record Optional. The event record. Uses the record from the RecordContext if not provided.
 * @param options.variant Optional. Either 'full', 'inline' or 'record'.
 * @param options.pluralizeResource Optional. A boolean indicating whether to pluralize the resource name.
 */
export const useEventLabel = (options: UseEventLabelOptions): string => {
    const { variant = 'full', pluralizeResource } = options;
    const record = useRecordContext<EventRecord>(options);
    const getResourceLabel = useGetResourceLabel();

    const translate = useTranslate();

    if (!record) {
        return null;
    }

    const pluralize =
        pluralizeResource != undefined
            ? pluralizeResource
            : actionPluralOptions[record.action];

    const { data = {}, ids = [] } = record.payload;
    const name =
        data.name || data.label || data.title || data.reference || data.id;

    const fields = [];
    if (record.payload.previousData) {
        for (const field of Object.keys(record.payload.previousData)) {
            if (
                record.payload.data[field] !==
                record.payload.previousData[field]
            ) {
                fields.push(
                    translate(
                        ...getFieldLabelTranslationArgs({
                            source: field,
                            resource: record.resource,
                        })
                    )
                );
            }
        }
    }

    const translateOptions = {
        name,
        ids: ids.join(', '),
        resource: getResourceLabel(
            record.resource,
            pluralize ? 2 : 1
        ).toLowerCase(),
        fields: fields.slice(0, 2).join(', '),
        all_fields: fields.join(', '),
        smart_count:
            variant === 'record' && fields.length > 2
                ? fields.length - 2
                : fields.length,
    };

    const finalVariant =
        variant === 'record' && fields.length > 2
            ? 'record_many_fields'
            : variant;

    const label = translate(
        `ra-audit-log.${variantTranslationKeys[finalVariant]}.${record.resource}.${record.action}`,
        {
            ...translateOptions,
            // Default generic translation when there is no resource specific translation available
            _: translate(
                `ra-audit-log.${variantTranslationKeys[finalVariant]}.${record.action}`,
                translateOptions
            ),
        }
    );

    return label;
};

const variantTranslationKeys = {
    full: 'event',
    inline: 'inline_event',
    record: 'record_event',
    record_many_fields: 'record_many_fields_events',
};

const actionPluralOptions = {
    create: false,
    delete: false,
    deleteMany: true,
    update: false,
    updateMany: true,
};

export interface UseEventLabelOptions {
    record?: EventRecord;
    pluralizeResource?: boolean;
    variant?: 'full' | 'inline' | 'record' | 'record_many_fields';
}
