// @react-admin/ra-realtime/lib/src/Warning.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Warning = void 0;
var react_1 = __importDefault(require("react"));
var react_admin_1 = require("react-admin");
var core_1 = require("@material-ui/core");
var styles_1 = require("@material-ui/core/styles");
var Warning = function (props) {
    var message = props.message, refresh = props.refresh, onRefresh = props.onRefresh;
    var classes = useStyles(props);
    var translate = react_admin_1.useTranslate();
    return (react_1.default.createElement(core_1.Paper, { className: classes.root, role: "alert", square: true, elevation: 0 },
        react_1.default.createElement("div", { className: classes.message },
            react_1.default.createElement(core_1.Typography, { gutterBottom: true, component: "div" }, translate('ra-realtime.notification.title')),
            message),
        react_1.default.createElement("div", { className: classes.action }, refresh && react_1.default.createElement(react_admin_1.RefreshButton, { onClick: onRefresh }))));
};
exports.Warning = Warning;
var useStyles = styles_1.makeStyles(function (theme) {
    var getColor = theme.palette.type === 'light' ? styles_1.darken : styles_1.lighten;
    var getBackgroundColor = theme.palette.type === 'light' ? styles_1.lighten : styles_1.darken;
    return {
        root: __assign(__assign({}, theme.typography.body2), { borderRadius: theme.shape.borderRadius, display: 'flex', padding: '6px 16px', margin: theme.spacing(1) + "px 0", color: getColor(theme.palette.warning.main, 0.6), backgroundColor: getBackgroundColor(theme.palette.warning.main, 0.9) }),
        message: {
            padding: theme.spacing(1) + "px 0",
        },
        action: {
            display: 'flex',
            alignItems: 'center',
            marginLeft: 'auto',
            paddingLeft: theme.spacing(2),
            marginRight: -theme.spacing(1),
        },
    };
}, {
    name: 'RaWarning',
});
